<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="author" content="L Collado-Torres" />

<meta name="date" content="2016-02-21" />

<title>Evaluate results</title>


<style type="text/css">code{white-space: pre;}</style>
<link href="data:text/css,pre%20%2Eoperator%2C%0Apre%20%2Eparen%20%7B%0A%20color%3A%20rgb%28104%2C%20118%2C%20135%29%0A%7D%0A%0Apre%20%2Eliteral%20%7B%0A%20color%3A%20%23990073%0A%7D%0A%0Apre%20%2Enumber%20%7B%0A%20color%3A%20%23099%3B%0A%7D%0A%0Apre%20%2Ecomment%20%7B%0A%20color%3A%20%23998%3B%0A%20font%2Dstyle%3A%20italic%0A%7D%0A%0Apre%20%2Ekeyword%20%7B%0A%20color%3A%20%23900%3B%0A%20font%2Dweight%3A%20bold%0A%7D%0A%0Apre%20%2Eidentifier%20%7B%0A%20color%3A%20rgb%280%2C%200%2C%200%29%3B%0A%7D%0A%0Apre%20%2Estring%20%7B%0A%20color%3A%20%23d14%3B%0A%7D%0A" rel="stylesheet" type="text/css" />
<script src="data:application/x-javascript,%0Avar%20hljs%3Dnew%20function%28%29%7Bfunction%20m%28p%29%7Breturn%20p%2Ereplace%28%2F%26%2Fgm%2C%22%26amp%3B%22%29%2Ereplace%28%2F%3C%2Fgm%2C%22%26lt%3B%22%29%7Dfunction%20f%28r%2Cq%2Cp%29%7Breturn%20RegExp%28q%2C%22m%22%2B%28r%2EcI%3F%22i%22%3A%22%22%29%2B%28p%3F%22g%22%3A%22%22%29%29%7Dfunction%20b%28r%29%7Bfor%28var%20p%3D0%3Bp%3Cr%2EchildNodes%2Elength%3Bp%2B%2B%29%7Bvar%20q%3Dr%2EchildNodes%5Bp%5D%3Bif%28q%2EnodeName%3D%3D%22CODE%22%29%7Breturn%20q%7Dif%28%21%28q%2EnodeType%3D%3D3%26%26q%2EnodeValue%2Ematch%28%2F%5Cs%2B%2F%29%29%29%7Bbreak%7D%7D%7Dfunction%20h%28t%2Cs%29%7Bvar%20p%3D%22%22%3Bfor%28var%20r%3D0%3Br%3Ct%2EchildNodes%2Elength%3Br%2B%2B%29%7Bif%28t%2EchildNodes%5Br%5D%2EnodeType%3D%3D3%29%7Bvar%20q%3Dt%2EchildNodes%5Br%5D%2EnodeValue%3Bif%28s%29%7Bq%3Dq%2Ereplace%28%2F%5Cn%2Fg%2C%22%22%29%7Dp%2B%3Dq%7Delse%7Bif%28t%2EchildNodes%5Br%5D%2EnodeName%3D%3D%22BR%22%29%7Bp%2B%3D%22%5Cn%22%7Delse%7Bp%2B%3Dh%28t%2EchildNodes%5Br%5D%29%7D%7D%7Dif%28%2FMSIE%20%5B678%5D%2F%2Etest%28navigator%2EuserAgent%29%29%7Bp%3Dp%2Ereplace%28%2F%5Cr%2Fg%2C%22%5Cn%22%29%7Dreturn%20p%7Dfunction%20a%28s%29%7Bvar%20r%3Ds%2EclassName%2Esplit%28%2F%5Cs%2B%2F%29%3Br%3Dr%2Econcat%28s%2EparentNode%2EclassName%2Esplit%28%2F%5Cs%2B%2F%29%29%3Bfor%28var%20q%3D0%3Bq%3Cr%2Elength%3Bq%2B%2B%29%7Bvar%20p%3Dr%5Bq%5D%2Ereplace%28%2F%5Elanguage%2D%2F%2C%22%22%29%3Bif%28e%5Bp%5D%29%7Breturn%20p%7D%7D%7Dfunction%20c%28q%29%7Bvar%20p%3D%5B%5D%3B%28function%28s%2Ct%29%7Bfor%28var%20r%3D0%3Br%3Cs%2EchildNodes%2Elength%3Br%2B%2B%29%7Bif%28s%2EchildNodes%5Br%5D%2EnodeType%3D%3D3%29%7Bt%2B%3Ds%2EchildNodes%5Br%5D%2EnodeValue%2Elength%7Delse%7Bif%28s%2EchildNodes%5Br%5D%2EnodeName%3D%3D%22BR%22%29%7Bt%2B%3D1%7Delse%7Bif%28s%2EchildNodes%5Br%5D%2EnodeType%3D%3D1%29%7Bp%2Epush%28%7Bevent%3A%22start%22%2Coffset%3At%2Cnode%3As%2EchildNodes%5Br%5D%7D%29%3Bt%3Darguments%2Ecallee%28s%2EchildNodes%5Br%5D%2Ct%29%3Bp%2Epush%28%7Bevent%3A%22stop%22%2Coffset%3At%2Cnode%3As%2EchildNodes%5Br%5D%7D%29%7D%7D%7D%7Dreturn%20t%7D%29%28q%2C0%29%3Breturn%20p%7Dfunction%20k%28y%2Cw%2Cx%29%7Bvar%20q%3D0%3Bvar%20z%3D%22%22%3Bvar%20s%3D%5B%5D%3Bfunction%20u%28%29%7Bif%28y%2Elength%26%26w%2Elength%29%7Bif%28y%5B0%5D%2Eoffset%21%3Dw%5B0%5D%2Eoffset%29%7Breturn%28y%5B0%5D%2Eoffset%3Cw%5B0%5D%2Eoffset%29%3Fy%3Aw%7Delse%7Breturn%20w%5B0%5D%2Eevent%3D%3D%22start%22%3Fy%3Aw%7D%7Delse%7Breturn%20y%2Elength%3Fy%3Aw%7D%7Dfunction%20t%28D%29%7Bvar%20A%3D%22%3C%22%2BD%2EnodeName%2EtoLowerCase%28%29%3Bfor%28var%20B%3D0%3BB%3CD%2Eattributes%2Elength%3BB%2B%2B%29%7Bvar%20C%3DD%2Eattributes%5BB%5D%3BA%2B%3D%22%20%22%2BC%2EnodeName%2EtoLowerCase%28%29%3Bif%28C%2Evalue%21%3D%3Dundefined%26%26C%2Evalue%21%3D%3Dfalse%26%26C%2Evalue%21%3D%3Dnull%29%7BA%2B%3D%27%3D%22%27%2Bm%28C%2Evalue%29%2B%27%22%27%7D%7Dreturn%20A%2B%22%3E%22%7Dwhile%28y%2Elength%7C%7Cw%2Elength%29%7Bvar%20v%3Du%28%29%2Esplice%280%2C1%29%5B0%5D%3Bz%2B%3Dm%28x%2Esubstr%28q%2Cv%2Eoffset%2Dq%29%29%3Bq%3Dv%2Eoffset%3Bif%28v%2Eevent%3D%3D%22start%22%29%7Bz%2B%3Dt%28v%2Enode%29%3Bs%2Epush%28v%2Enode%29%7Delse%7Bif%28v%2Eevent%3D%3D%22stop%22%29%7Bvar%20p%2Cr%3Ds%2Elength%3Bdo%7Br%2D%2D%3Bp%3Ds%5Br%5D%3Bz%2B%3D%28%22%3C%2F%22%2Bp%2EnodeName%2EtoLowerCase%28%29%2B%22%3E%22%29%7Dwhile%28p%21%3Dv%2Enode%29%3Bs%2Esplice%28r%2C1%29%3Bwhile%28r%3Cs%2Elength%29%7Bz%2B%3Dt%28s%5Br%5D%29%3Br%2B%2B%7D%7D%7D%7Dreturn%20z%2Bm%28x%2Esubstr%28q%29%29%7Dfunction%20j%28%29%7Bfunction%20q%28x%2Cy%2Cv%29%7Bif%28x%2Ecompiled%29%7Breturn%7Dvar%20u%3Bvar%20s%3D%5B%5D%3Bif%28x%2Ek%29%7Bx%2ElR%3Df%28y%2Cx%2El%7C%7Chljs%2EIR%2Ctrue%29%3Bfor%28var%20w%20in%20x%2Ek%29%7Bif%28%21x%2Ek%2EhasOwnProperty%28w%29%29%7Bcontinue%7Dif%28x%2Ek%5Bw%5D%20instanceof%20Object%29%7Bu%3Dx%2Ek%5Bw%5D%7Delse%7Bu%3Dx%2Ek%3Bw%3D%22keyword%22%7Dfor%28var%20r%20in%20u%29%7Bif%28%21u%2EhasOwnProperty%28r%29%29%7Bcontinue%7Dx%2Ek%5Br%5D%3D%5Bw%2Cu%5Br%5D%5D%3Bs%2Epush%28r%29%7D%7D%7Dif%28%21v%29%7Bif%28x%2EbWK%29%7Bx%2Eb%3D%22%5C%5Cb%28%22%2Bs%2Ejoin%28%22%7C%22%29%2B%22%29%5C%5Cs%22%7Dx%2EbR%3Df%28y%2Cx%2Eb%3Fx%2Eb%3A%22%5C%5CB%7C%5C%5Cb%22%29%3Bif%28%21x%2Ee%26%26%21x%2EeW%29%7Bx%2Ee%3D%22%5C%5CB%7C%5C%5Cb%22%7Dif%28x%2Ee%29%7Bx%2EeR%3Df%28y%2Cx%2Ee%29%7D%7Dif%28x%2Ei%29%7Bx%2EiR%3Df%28y%2Cx%2Ei%29%7Dif%28x%2Er%3D%3D%3Dundefined%29%7Bx%2Er%3D1%7Dif%28%21x%2Ec%29%7Bx%2Ec%3D%5B%5D%7Dx%2Ecompiled%3Dtrue%3Bfor%28var%20t%3D0%3Bt%3Cx%2Ec%2Elength%3Bt%2B%2B%29%7Bif%28x%2Ec%5Bt%5D%3D%3D%22self%22%29%7Bx%2Ec%5Bt%5D%3Dx%7Dq%28x%2Ec%5Bt%5D%2Cy%2Cfalse%29%7Dif%28x%2Estarts%29%7Bq%28x%2Estarts%2Cy%2Cfalse%29%7D%7Dfor%28var%20p%20in%20e%29%7Bif%28%21e%2EhasOwnProperty%28p%29%29%7Bcontinue%7Dq%28e%5Bp%5D%2EdM%2Ce%5Bp%5D%2Ctrue%29%7D%7Dfunction%20d%28B%2CC%29%7Bif%28%21j%2Ecalled%29%7Bj%28%29%3Bj%2Ecalled%3Dtrue%7Dfunction%20q%28r%2CM%29%7Bfor%28var%20L%3D0%3BL%3CM%2Ec%2Elength%3BL%2B%2B%29%7Bif%28%28M%2Ec%5BL%5D%2EbR%2Eexec%28r%29%7C%7C%5Bnull%5D%29%5B0%5D%3D%3Dr%29%7Breturn%20M%2Ec%5BL%5D%7D%7D%7Dfunction%20v%28L%2Cr%29%7Bif%28D%5BL%5D%2Ee%26%26D%5BL%5D%2EeR%2Etest%28r%29%29%7Breturn%201%7Dif%28D%5BL%5D%2EeW%29%7Bvar%20M%3Dv%28L%2D1%2Cr%29%3Breturn%20M%3FM%2B1%3A0%7Dreturn%200%7Dfunction%20w%28r%2CL%29%7Breturn%20L%2Ei%26%26L%2EiR%2Etest%28r%29%7Dfunction%20K%28N%2CO%29%7Bvar%20M%3D%5B%5D%3Bfor%28var%20L%3D0%3BL%3CN%2Ec%2Elength%3BL%2B%2B%29%7BM%2Epush%28N%2Ec%5BL%5D%2Eb%29%7Dvar%20r%3DD%2Elength%2D1%3Bdo%7Bif%28D%5Br%5D%2Ee%29%7BM%2Epush%28D%5Br%5D%2Ee%29%7Dr%2D%2D%7Dwhile%28D%5Br%2B1%5D%2EeW%29%3Bif%28N%2Ei%29%7BM%2Epush%28N%2Ei%29%7Dreturn%20f%28O%2CM%2Ejoin%28%22%7C%22%29%2Ctrue%29%7Dfunction%20p%28M%2CL%29%7Bvar%20N%3DD%5BD%2Elength%2D1%5D%3Bif%28%21N%2Et%29%7BN%2Et%3DK%28N%2CE%29%7DN%2Et%2ElastIndex%3DL%3Bvar%20r%3DN%2Et%2Eexec%28M%29%3Breturn%20r%3F%5BM%2Esubstr%28L%2Cr%2Eindex%2DL%29%2Cr%5B0%5D%2Cfalse%5D%3A%5BM%2Esubstr%28L%29%2C%22%22%2Ctrue%5D%7Dfunction%20z%28N%2Cr%29%7Bvar%20L%3DE%2EcI%3Fr%5B0%5D%2EtoLowerCase%28%29%3Ar%5B0%5D%3Bvar%20M%3DN%2Ek%5BL%5D%3Bif%28M%26%26M%20instanceof%20Array%29%7Breturn%20M%7Dreturn%20false%7Dfunction%20F%28L%2CP%29%7BL%3Dm%28L%29%3Bif%28%21P%2Ek%29%7Breturn%20L%7Dvar%20r%3D%22%22%3Bvar%20O%3D0%3BP%2ElR%2ElastIndex%3D0%3Bvar%20M%3DP%2ElR%2Eexec%28L%29%3Bwhile%28M%29%7Br%2B%3DL%2Esubstr%28O%2CM%2Eindex%2DO%29%3Bvar%20N%3Dz%28P%2CM%29%3Bif%28N%29%7Bx%2B%3DN%5B1%5D%3Br%2B%3D%27%3Cspan%20class%3D%22%27%2BN%5B0%5D%2B%27%22%3E%27%2BM%5B0%5D%2B%22%3C%2Fspan%3E%22%7Delse%7Br%2B%3DM%5B0%5D%7DO%3DP%2ElR%2ElastIndex%3BM%3DP%2ElR%2Eexec%28L%29%7Dreturn%20r%2BL%2Esubstr%28O%2CL%2Elength%2DO%29%7Dfunction%20J%28L%2CM%29%7Bif%28M%2EsL%26%26e%5BM%2EsL%5D%29%7Bvar%20r%3Dd%28M%2EsL%2CL%29%3Bx%2B%3Dr%2Ekeyword%5Fcount%3Breturn%20r%2Evalue%7Delse%7Breturn%20F%28L%2CM%29%7D%7Dfunction%20I%28M%2Cr%29%7Bvar%20L%3DM%2EcN%3F%27%3Cspan%20class%3D%22%27%2BM%2EcN%2B%27%22%3E%27%3A%22%22%3Bif%28M%2ErB%29%7By%2B%3DL%3BM%2Ebuffer%3D%22%22%7Delse%7Bif%28M%2EeB%29%7By%2B%3Dm%28r%29%2BL%3BM%2Ebuffer%3D%22%22%7Delse%7By%2B%3DL%3BM%2Ebuffer%3Dr%7D%7DD%2Epush%28M%29%3BA%2B%3DM%2Er%7Dfunction%20G%28N%2CM%2CQ%29%7Bvar%20R%3DD%5BD%2Elength%2D1%5D%3Bif%28Q%29%7By%2B%3DJ%28R%2Ebuffer%2BN%2CR%29%3Breturn%20false%7Dvar%20P%3Dq%28M%2CR%29%3Bif%28P%29%7By%2B%3DJ%28R%2Ebuffer%2BN%2CR%29%3BI%28P%2CM%29%3Breturn%20P%2ErB%7Dvar%20L%3Dv%28D%2Elength%2D1%2CM%29%3Bif%28L%29%7Bvar%20O%3DR%2EcN%3F%22%3C%2Fspan%3E%22%3A%22%22%3Bif%28R%2ErE%29%7By%2B%3DJ%28R%2Ebuffer%2BN%2CR%29%2BO%7Delse%7Bif%28R%2EeE%29%7By%2B%3DJ%28R%2Ebuffer%2BN%2CR%29%2BO%2Bm%28M%29%7Delse%7By%2B%3DJ%28R%2Ebuffer%2BN%2BM%2CR%29%2BO%7D%7Dwhile%28L%3E1%29%7BO%3DD%5BD%2Elength%2D2%5D%2EcN%3F%22%3C%2Fspan%3E%22%3A%22%22%3By%2B%3DO%3BL%2D%2D%3BD%2Elength%2D%2D%7Dvar%20r%3DD%5BD%2Elength%2D1%5D%3BD%2Elength%2D%2D%3BD%5BD%2Elength%2D1%5D%2Ebuffer%3D%22%22%3Bif%28r%2Estarts%29%7BI%28r%2Estarts%2C%22%22%29%7Dreturn%20R%2ErE%7Dif%28w%28M%2CR%29%29%7Bthrow%22Illegal%22%7D%7Dvar%20E%3De%5BB%5D%3Bvar%20D%3D%5BE%2EdM%5D%3Bvar%20A%3D0%3Bvar%20x%3D0%3Bvar%20y%3D%22%22%3Btry%7Bvar%20s%2Cu%3D0%3BE%2EdM%2Ebuffer%3D%22%22%3Bdo%7Bs%3Dp%28C%2Cu%29%3Bvar%20t%3DG%28s%5B0%5D%2Cs%5B1%5D%2Cs%5B2%5D%29%3Bu%2B%3Ds%5B0%5D%2Elength%3Bif%28%21t%29%7Bu%2B%3Ds%5B1%5D%2Elength%7D%7Dwhile%28%21s%5B2%5D%29%3Bif%28D%2Elength%3E1%29%7Bthrow%22Illegal%22%7Dreturn%7Br%3AA%2Ckeyword%5Fcount%3Ax%2Cvalue%3Ay%7D%7Dcatch%28H%29%7Bif%28H%3D%3D%22Illegal%22%29%7Breturn%7Br%3A0%2Ckeyword%5Fcount%3A0%2Cvalue%3Am%28C%29%7D%7Delse%7Bthrow%20H%7D%7D%7Dfunction%20g%28t%29%7Bvar%20p%3D%7Bkeyword%5Fcount%3A0%2Cr%3A0%2Cvalue%3Am%28t%29%7D%3Bvar%20r%3Dp%3Bfor%28var%20q%20in%20e%29%7Bif%28%21e%2EhasOwnProperty%28q%29%29%7Bcontinue%7Dvar%20s%3Dd%28q%2Ct%29%3Bs%2Elanguage%3Dq%3Bif%28s%2Ekeyword%5Fcount%2Bs%2Er%3Er%2Ekeyword%5Fcount%2Br%2Er%29%7Br%3Ds%7Dif%28s%2Ekeyword%5Fcount%2Bs%2Er%3Ep%2Ekeyword%5Fcount%2Bp%2Er%29%7Br%3Dp%3Bp%3Ds%7D%7Dif%28r%2Elanguage%29%7Bp%2Esecond%5Fbest%3Dr%7Dreturn%20p%7Dfunction%20i%28r%2Cq%2Cp%29%7Bif%28q%29%7Br%3Dr%2Ereplace%28%2F%5E%28%28%3C%5B%5E%3E%5D%2B%3E%7C%5Ct%29%2B%29%2Fgm%2Cfunction%28t%2Cw%2Cv%2Cu%29%7Breturn%20w%2Ereplace%28%2F%5Ct%2Fg%2Cq%29%7D%29%7Dif%28p%29%7Br%3Dr%2Ereplace%28%2F%5Cn%2Fg%2C%22%3Cbr%3E%22%29%7Dreturn%20r%7Dfunction%20n%28t%2Cw%2Cr%29%7Bvar%20x%3Dh%28t%2Cr%29%3Bvar%20v%3Da%28t%29%3Bvar%20y%2Cs%3Bif%28v%29%7By%3Dd%28v%2Cx%29%7Delse%7Breturn%7Dvar%20q%3Dc%28t%29%3Bif%28q%2Elength%29%7Bs%3Ddocument%2EcreateElement%28%22pre%22%29%3Bs%2EinnerHTML%3Dy%2Evalue%3By%2Evalue%3Dk%28q%2Cc%28s%29%2Cx%29%7Dy%2Evalue%3Di%28y%2Evalue%2Cw%2Cr%29%3Bvar%20u%3Dt%2EclassName%3Bif%28%21u%2Ematch%28%22%28%5C%5Cs%7C%5E%29%28language%2D%29%3F%22%2Bv%2B%22%28%5C%5Cs%7C%24%29%22%29%29%7Bu%3Du%3F%28u%2B%22%20%22%2Bv%29%3Av%7Dif%28%2FMSIE%20%5B678%5D%2F%2Etest%28navigator%2EuserAgent%29%26%26t%2EtagName%3D%3D%22CODE%22%26%26t%2EparentNode%2EtagName%3D%3D%22PRE%22%29%7Bs%3Dt%2EparentNode%3Bvar%20p%3Ddocument%2EcreateElement%28%22div%22%29%3Bp%2EinnerHTML%3D%22%3Cpre%3E%3Ccode%3E%22%2By%2Evalue%2B%22%3C%2Fcode%3E%3C%2Fpre%3E%22%3Bt%3Dp%2EfirstChild%2EfirstChild%3Bp%2EfirstChild%2EcN%3Ds%2EcN%3Bs%2EparentNode%2EreplaceChild%28p%2EfirstChild%2Cs%29%7Delse%7Bt%2EinnerHTML%3Dy%2Evalue%7Dt%2EclassName%3Du%3Bt%2Eresult%3D%7Blanguage%3Av%2Ckw%3Ay%2Ekeyword%5Fcount%2Cre%3Ay%2Er%7D%3Bif%28y%2Esecond%5Fbest%29%7Bt%2Esecond%5Fbest%3D%7Blanguage%3Ay%2Esecond%5Fbest%2Elanguage%2Ckw%3Ay%2Esecond%5Fbest%2Ekeyword%5Fcount%2Cre%3Ay%2Esecond%5Fbest%2Er%7D%7D%7Dfunction%20o%28%29%7Bif%28o%2Ecalled%29%7Breturn%7Do%2Ecalled%3Dtrue%3Bvar%20r%3Ddocument%2EgetElementsByTagName%28%22pre%22%29%3Bfor%28var%20p%3D0%3Bp%3Cr%2Elength%3Bp%2B%2B%29%7Bvar%20q%3Db%28r%5Bp%5D%29%3Bif%28q%29%7Bn%28q%2Chljs%2EtabReplace%29%7D%7D%7Dfunction%20l%28%29%7Bif%28window%2EaddEventListener%29%7Bwindow%2EaddEventListener%28%22DOMContentLoaded%22%2Co%2Cfalse%29%3Bwindow%2EaddEventListener%28%22load%22%2Co%2Cfalse%29%7Delse%7Bif%28window%2EattachEvent%29%7Bwindow%2EattachEvent%28%22onload%22%2Co%29%7Delse%7Bwindow%2Eonload%3Do%7D%7D%7Dvar%20e%3D%7B%7D%3Bthis%2ELANGUAGES%3De%3Bthis%2Ehighlight%3Dd%3Bthis%2EhighlightAuto%3Dg%3Bthis%2EfixMarkup%3Di%3Bthis%2EhighlightBlock%3Dn%3Bthis%2EinitHighlighting%3Do%3Bthis%2EinitHighlightingOnLoad%3Dl%3Bthis%2EIR%3D%22%5Ba%2DzA%2DZ%5D%5Ba%2DzA%2DZ0%2D9%5F%5D%2A%22%3Bthis%2EUIR%3D%22%5Ba%2DzA%2DZ%5F%5D%5Ba%2DzA%2DZ0%2D9%5F%5D%2A%22%3Bthis%2ENR%3D%22%5C%5Cb%5C%5Cd%2B%28%5C%5C%2E%5C%5Cd%2B%29%3F%22%3Bthis%2ECNR%3D%22%5C%5Cb%280%5BxX%5D%5Ba%2DfA%2DF0%2D9%5D%2B%7C%28%5C%5Cd%2B%28%5C%5C%2E%5C%5Cd%2A%29%3F%7C%5C%5C%2E%5C%5Cd%2B%29%28%5BeE%5D%5B%2D%2B%5D%3F%5C%5Cd%2B%29%3F%29%22%3Bthis%2EBNR%3D%22%5C%5Cb%280b%5B01%5D%2B%29%22%3Bthis%2ERSR%3D%22%21%7C%21%3D%7C%21%3D%3D%7C%25%7C%25%3D%7C%26%7C%26%26%7C%26%3D%7C%5C%5C%2A%7C%5C%5C%2A%3D%7C%5C%5C%2B%7C%5C%5C%2B%3D%7C%2C%7C%5C%5C%2E%7C%2D%7C%2D%3D%7C%2F%7C%2F%3D%7C%3A%7C%3B%7C%3C%7C%3C%3C%7C%3C%3C%3D%7C%3C%3D%7C%3D%7C%3D%3D%7C%3D%3D%3D%7C%3E%7C%3E%3D%7C%3E%3E%7C%3E%3E%3D%7C%3E%3E%3E%7C%3E%3E%3E%3D%7C%5C%5C%3F%7C%5C%5C%5B%7C%5C%5C%7B%7C%5C%5C%28%7C%5C%5C%5E%7C%5C%5C%5E%3D%7C%5C%5C%7C%7C%5C%5C%7C%3D%7C%5C%5C%7C%5C%5C%7C%7C%7E%22%3Bthis%2EER%3D%22%28%3F%21%5B%5C%5Cs%5C%5CS%5D%29%22%3Bthis%2EBE%3D%7Bb%3A%22%5C%5C%5C%5C%2E%22%2Cr%3A0%7D%3Bthis%2EASM%3D%7BcN%3A%22string%22%2Cb%3A%22%27%22%2Ce%3A%22%27%22%2Ci%3A%22%5C%5Cn%22%2Cc%3A%5Bthis%2EBE%5D%2Cr%3A0%7D%3Bthis%2EQSM%3D%7BcN%3A%22string%22%2Cb%3A%27%22%27%2Ce%3A%27%22%27%2Ci%3A%22%5C%5Cn%22%2Cc%3A%5Bthis%2EBE%5D%2Cr%3A0%7D%3Bthis%2ECLCM%3D%7BcN%3A%22comment%22%2Cb%3A%22%2F%2F%22%2Ce%3A%22%24%22%7D%3Bthis%2ECBLCLM%3D%7BcN%3A%22comment%22%2Cb%3A%22%2F%5C%5C%2A%22%2Ce%3A%22%5C%5C%2A%2F%22%7D%3Bthis%2EHCM%3D%7BcN%3A%22comment%22%2Cb%3A%22%23%22%2Ce%3A%22%24%22%7D%3Bthis%2ENM%3D%7BcN%3A%22number%22%2Cb%3Athis%2ENR%2Cr%3A0%7D%3Bthis%2ECNM%3D%7BcN%3A%22number%22%2Cb%3Athis%2ECNR%2Cr%3A0%7D%3Bthis%2EBNM%3D%7BcN%3A%22number%22%2Cb%3Athis%2EBNR%2Cr%3A0%7D%3Bthis%2Einherit%3Dfunction%28r%2Cs%29%7Bvar%20p%3D%7B%7D%3Bfor%28var%20q%20in%20r%29%7Bp%5Bq%5D%3Dr%5Bq%5D%7Dif%28s%29%7Bfor%28var%20q%20in%20s%29%7Bp%5Bq%5D%3Ds%5Bq%5D%7D%7Dreturn%20p%7D%7D%28%29%3Bhljs%2ELANGUAGES%2Ecpp%3Dfunction%28%29%7Bvar%20a%3D%7Bkeyword%3A%7B%22false%22%3A1%2C%22int%22%3A1%2C%22float%22%3A1%2C%22while%22%3A1%2C%22private%22%3A1%2C%22char%22%3A1%2C%22catch%22%3A1%2C%22export%22%3A1%2Cvirtual%3A1%2Coperator%3A2%2Csizeof%3A2%2Cdynamic%5Fcast%3A2%2Ctypedef%3A2%2Cconst%5Fcast%3A2%2C%22const%22%3A1%2Cstruct%3A1%2C%22for%22%3A1%2Cstatic%5Fcast%3A2%2Cunion%3A1%2Cnamespace%3A1%2Cunsigned%3A1%2C%22long%22%3A1%2C%22throw%22%3A1%2C%22volatile%22%3A2%2C%22static%22%3A1%2C%22protected%22%3A1%2Cbool%3A1%2Ctemplate%3A1%2Cmutable%3A1%2C%22if%22%3A1%2C%22public%22%3A1%2Cfriend%3A2%2C%22do%22%3A1%2C%22return%22%3A1%2C%22goto%22%3A1%2Cauto%3A1%2C%22void%22%3A2%2C%22enum%22%3A1%2C%22else%22%3A1%2C%22break%22%3A1%2C%22new%22%3A1%2Cextern%3A1%2Cusing%3A1%2C%22true%22%3A1%2C%22class%22%3A1%2Casm%3A1%2C%22case%22%3A1%2Ctypeid%3A1%2C%22short%22%3A1%2Creinterpret%5Fcast%3A2%2C%22default%22%3A1%2C%22double%22%3A1%2Cregister%3A1%2Cexplicit%3A1%2Csigned%3A1%2Ctypename%3A1%2C%22try%22%3A1%2C%22this%22%3A1%2C%22switch%22%3A1%2C%22continue%22%3A1%2Cwchar%5Ft%3A1%2Cinline%3A1%2C%22delete%22%3A1%2Calignof%3A1%2Cchar16%5Ft%3A1%2Cchar32%5Ft%3A1%2Cconstexpr%3A1%2Cdecltype%3A1%2Cnoexcept%3A1%2Cnullptr%3A1%2Cstatic%5Fassert%3A1%2Cthread%5Flocal%3A1%2Crestrict%3A1%2C%5FBool%3A1%2Ccomplex%3A1%7D%2Cbuilt%5Fin%3A%7Bstd%3A1%2Cstring%3A1%2Ccin%3A1%2Ccout%3A1%2Ccerr%3A1%2Cclog%3A1%2Cstringstream%3A1%2Cistringstream%3A1%2Costringstream%3A1%2Cauto%5Fptr%3A1%2Cdeque%3A1%2Clist%3A1%2Cqueue%3A1%2Cstack%3A1%2Cvector%3A1%2Cmap%3A1%2Cset%3A1%2Cbitset%3A1%2Cmultiset%3A1%2Cmultimap%3A1%2Cunordered%5Fset%3A1%2Cunordered%5Fmap%3A1%2Cunordered%5Fmultiset%3A1%2Cunordered%5Fmultimap%3A1%2Carray%3A1%2Cshared%5Fptr%3A1%7D%7D%3Breturn%7BdM%3A%7Bk%3Aa%2Ci%3A%22%3C%2F%22%2Cc%3A%5Bhljs%2ECLCM%2Chljs%2ECBLCLM%2Chljs%2EQSM%2C%7BcN%3A%22string%22%2Cb%3A%22%27%5C%5C%5C%5C%3F%2E%22%2Ce%3A%22%27%22%2Ci%3A%22%2E%22%7D%2C%7BcN%3A%22number%22%2Cb%3A%22%5C%5Cb%28%5C%5Cd%2B%28%5C%5C%2E%5C%5Cd%2A%29%3F%7C%5C%5C%2E%5C%5Cd%2B%29%28u%7CU%7Cl%7CL%7Cul%7CUL%7Cf%7CF%29%22%7D%2Chljs%2ECNM%2C%7BcN%3A%22preprocessor%22%2Cb%3A%22%23%22%2Ce%3A%22%24%22%7D%2C%7BcN%3A%22stl%5Fcontainer%22%2Cb%3A%22%5C%5Cb%28deque%7Clist%7Cqueue%7Cstack%7Cvector%7Cmap%7Cset%7Cbitset%7Cmultiset%7Cmultimap%7Cunordered%5Fmap%7Cunordered%5Fset%7Cunordered%5Fmultiset%7Cunordered%5Fmultimap%7Carray%29%5C%5Cs%2A%3C%22%2Ce%3A%22%3E%22%2Ck%3Aa%2Cr%3A10%2Cc%3A%5B%22self%22%5D%7D%5D%7D%7D%7D%28%29%3Bhljs%2ELANGUAGES%2Er%3D%7BdM%3A%7Bc%3A%5Bhljs%2EHCM%2C%7BcN%3A%22number%22%2Cb%3A%22%5C%5Cb0%5BxX%5D%5B0%2D9a%2DfA%2DF%5D%2B%5BLi%5D%3F%5C%5Cb%22%2Ce%3Ahljs%2EIMMEDIATE%5FRE%2Cr%3A0%7D%2C%7BcN%3A%22number%22%2Cb%3A%22%5C%5Cb%5C%5Cd%2B%28%3F%3A%5BeE%5D%5B%2B%5C%5C%2D%5D%3F%5C%5Cd%2A%29%3FL%5C%5Cb%22%2Ce%3Ahljs%2EIMMEDIATE%5FRE%2Cr%3A0%7D%2C%7BcN%3A%22number%22%2Cb%3A%22%5C%5Cb%5C%5Cd%2B%5C%5C%2E%28%3F%21%5C%5Cd%29%28%3F%3Ai%5C%5Cb%29%3F%22%2Ce%3Ahljs%2EIMMEDIATE%5FRE%2Cr%3A1%7D%2C%7BcN%3A%22number%22%2Cb%3A%22%5C%5Cb%5C%5Cd%2B%28%3F%3A%5C%5C%2E%5C%5Cd%2A%29%3F%28%3F%3A%5BeE%5D%5B%2B%5C%5C%2D%5D%3F%5C%5Cd%2A%29%3Fi%3F%5C%5Cb%22%2Ce%3Ahljs%2EIMMEDIATE%5FRE%2Cr%3A0%7D%2C%7BcN%3A%22number%22%2Cb%3A%22%5C%5C%2E%5C%5Cd%2B%28%3F%3A%5BeE%5D%5B%2B%5C%5C%2D%5D%3F%5C%5Cd%2A%29%3Fi%3F%5C%5Cb%22%2Ce%3Ahljs%2EIMMEDIATE%5FRE%2Cr%3A1%7D%2C%7BcN%3A%22keyword%22%2Cb%3A%22%28%3F%3AtryCatch%7Clibrary%7CsetGeneric%7CsetGroupGeneric%29%5C%5Cb%22%2Ce%3Ahljs%2EIMMEDIATE%5FRE%2Cr%3A10%7D%2C%7BcN%3A%22keyword%22%2Cb%3A%22%5C%5C%2E%5C%5C%2E%5C%5C%2E%22%2Ce%3Ahljs%2EIMMEDIATE%5FRE%2Cr%3A10%7D%2C%7BcN%3A%22keyword%22%2Cb%3A%22%5C%5C%2E%5C%5C%2E%5C%5Cd%2B%28%3F%21%5B%5C%5Cw%2E%5D%29%22%2Ce%3Ahljs%2EIMMEDIATE%5FRE%2Cr%3A10%7D%2C%7BcN%3A%22keyword%22%2Cb%3A%22%5C%5Cb%28%3F%3Afunction%29%22%2Ce%3Ahljs%2EIMMEDIATE%5FRE%2Cr%3A2%7D%2C%7BcN%3A%22keyword%22%2Cb%3A%22%28%3F%3Aif%7Cin%7Cbreak%7Cnext%7Crepeat%7Celse%7Cfor%7Creturn%7Cswitch%7Cwhile%7Ctry%7Cstop%7Cwarning%7Crequire%7Cattach%7Cdetach%7Csource%7CsetMethod%7CsetClass%29%5C%5Cb%22%2Ce%3Ahljs%2EIMMEDIATE%5FRE%2Cr%3A1%7D%2C%7BcN%3A%22literal%22%2Cb%3A%22%28%3F%3ANA%7CNA%5Finteger%5F%7CNA%5Freal%5F%7CNA%5Fcharacter%5F%7CNA%5Fcomplex%5F%29%5C%5Cb%22%2Ce%3Ahljs%2EIMMEDIATE%5FRE%2Cr%3A10%7D%2C%7BcN%3A%22literal%22%2Cb%3A%22%28%3F%3ANULL%7CTRUE%7CFALSE%7CT%7CF%7CInf%7CNaN%29%5C%5Cb%22%2Ce%3Ahljs%2EIMMEDIATE%5FRE%2Cr%3A1%7D%2C%7BcN%3A%22identifier%22%2Cb%3A%22%5Ba%2DzA%2DZ%2E%5D%5Ba%2DzA%2DZ0%2D9%2E%5F%5D%2A%5C%5Cb%22%2Ce%3Ahljs%2EIMMEDIATE%5FRE%2Cr%3A0%7D%2C%7BcN%3A%22operator%22%2Cb%3A%22%3C%5C%5C%2D%28%3F%21%5C%5Cs%2A%5C%5Cd%29%22%2Ce%3Ahljs%2EIMMEDIATE%5FRE%2Cr%3A2%7D%2C%7BcN%3A%22operator%22%2Cb%3A%22%5C%5C%2D%3E%7C%3C%5C%5C%2D%22%2Ce%3Ahljs%2EIMMEDIATE%5FRE%2Cr%3A1%7D%2C%7BcN%3A%22operator%22%2Cb%3A%22%25%25%7C%7E%22%2Ce%3Ahljs%2EIMMEDIATE%5FRE%7D%2C%7BcN%3A%22operator%22%2Cb%3A%22%3E%3D%7C%3C%3D%7C%3D%3D%7C%21%3D%7C%5C%5C%7C%5C%5C%7C%7C%26%26%7C%3D%7C%5C%5C%2B%7C%5C%5C%2D%7C%5C%5C%2A%7C%2F%7C%5C%5C%5E%7C%3E%7C%3C%7C%21%7C%26%7C%5C%5C%7C%7C%5C%5C%24%7C%3A%22%2Ce%3Ahljs%2EIMMEDIATE%5FRE%2Cr%3A0%7D%2C%7BcN%3A%22operator%22%2Cb%3A%22%25%22%2Ce%3A%22%25%22%2Ci%3A%22%5C%5Cn%22%2Cr%3A1%7D%2C%7BcN%3A%22identifier%22%2Cb%3A%22%60%22%2Ce%3A%22%60%22%2Cr%3A0%7D%2C%7BcN%3A%22string%22%2Cb%3A%27%22%27%2Ce%3A%27%22%27%2Cc%3A%5Bhljs%2EBE%5D%2Cr%3A0%7D%2C%7BcN%3A%22string%22%2Cb%3A%22%27%22%2Ce%3A%22%27%22%2Cc%3A%5Bhljs%2EBE%5D%2Cr%3A0%7D%2C%7BcN%3A%22paren%22%2Cb%3A%22%5B%5B%28%7B%5C%5C%5D%29%7D%5D%22%2Ce%3Ahljs%2EIMMEDIATE%5FRE%2Cr%3A0%7D%5D%7D%7D%3B%0Ahljs%2EinitHighlightingOnLoad%28%29%3B%0A%0A"></script>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs && document.readyState && document.readyState === "complete") {
   window.setTimeout(function() {
      hljs.initHighlighting();
   }, 0);
}
</script>


<link href="data:text/css,body%2C%20td%20%7B%0A%20%20%20font%2Dfamily%3A%20sans%2Dserif%3B%0A%20%20%20background%2Dcolor%3A%20white%3B%0A%20%20%20font%2Dsize%3A%2013px%3B%0A%7D%0A%0Abody%20%7B%0A%20%20max%2Dwidth%3A%20800px%3B%0A%20%20margin%3A%200%20auto%3B%0A%20%20padding%3A%201em%201em%202em%3B%0A%20%20line%2Dheight%3A%2020px%3B%0A%7D%0A%0A%2F%2A%20Table%20of%20contents%20style%20%2A%2F%0A%0Adiv%23TOC%20li%20%7B%0A%20%20%20%20list%2Dstyle%3Anone%3B%0A%20%20%20%20background%2Dimage%3Anone%3B%0A%20%20%20%20background%2Drepeat%3Anone%3B%0A%20%20%20%20background%2Dposition%3A0%3B%0A%7D%0A%0A%2F%2A%20element%20spacing%20%2A%2F%0A%0Ap%2C%20pre%20%7B%20%0A%20%20margin%3A%200em%200em%201em%3B%0A%7D%0A%0A%2F%2A%20center%20images%20and%20tables%20%2A%2F%0Aimg%2C%20table%20%7B%0A%20%20margin%3A%200em%20auto%201em%3B%0A%7D%0A%0Ap%20%7B%0A%20%20text%2Dalign%3A%20justify%3B%0A%7D%0A%0Att%2C%20code%2C%20pre%20%7B%0A%20%20%20font%2Dfamily%3A%20%27DejaVu%20Sans%20Mono%27%2C%20%27Droid%20Sans%20Mono%27%2C%20%27Lucida%20Console%27%2C%20Consolas%2C%20Monaco%2C%20monospace%3B%0A%7D%0A%0Ah1%2C%20h2%2C%20h3%2C%20h4%2C%20h5%2C%20h6%20%7B%20%0A%20%20font%2Dfamily%3A%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0A%20%20margin%3A%201%2E2em%200em%200%2E6em%200em%3B%0A%20%20font%2Dweight%3A%20bold%3B%0A%7D%0A%0Ah1%2Etitle%20%7B%0A%20%20font%2Dsize%3A%20250%25%3B%0A%20%20font%2Dweight%3A%20normal%3B%0A%20%20color%3A%20%2387b13f%3B%0A%20%20line%2Dheight%3A%201%2E1em%3B%0A%20%20margin%2Dtop%3A%200px%3B%0A%20%20border%2Dbottom%3A%200px%3B%0A%7D%0A%0Ah1%20%7B%0A%20%20font%2Dsize%3A%20160%25%3B%0A%20%20font%2Dweight%3A%20normal%3B%0A%20%20line%2Dheight%3A%201%2E4em%3B%0A%20%20border%2Dbottom%3A%201px%20%231a81c2%20solid%3B%0A%7D%0A%0Ah2%20%7B%0A%20%20font%2Dsize%3A%20130%25%3B%20%20%0A%7D%0A%0Ah1%2C%20h2%2C%20h3%20%7B%0A%20%20color%3A%20%231a81c2%3B%0A%7D%0A%0Ah3%2C%20h4%2C%20h5%2C%20h6%20%7B%0A%20%20font%2Dsize%3A115%25%3B%0A%7D%20%2F%2A%20not%20expecting%20to%20dive%20deeper%20than%20four%20levels%20on%20a%20single%20page%20%2A%2F%0A%0A%2F%2A%20links%20are%20simply%20blue%2C%20hovering%20slightly%20less%20blue%20%2A%2F%0Aa%20%7B%20color%3A%20%231a81c2%3B%20%7D%0Aa%3Aactive%20%7B%20outline%3A%20none%3B%20%7D%0Aa%3Avisited%20%7B%20color%3A%20%231a81c2%3B%20%7D%0Aa%3Ahover%20%7B%20color%3A%20%234c94c2%3B%20%7D%0A%0Apre%2C%20img%20%7B%0A%20%20max%2Dwidth%3A%20100%25%3B%0A%20%20display%3A%20block%3B%0A%7D%0A%0Apre%20%7B%0A%20%20border%3A%200px%20none%3B%0A%20%20background%2Dcolor%3A%20%23F8F8F8%3B%0A%20%20white%2Dspace%3A%20pre%3B%0A%20%20overflow%2Dx%3A%20auto%3B%0A%7D%0A%0Apre%20code%20%7B%0A%20%20border%3A%201px%20%23aaa%20dashed%3B%0A%20%20background%2Dcolor%3A%20white%3B%0A%20%20display%3A%20block%3B%0A%20%20padding%3A%201em%3B%20%20%0A%20%20color%3A%20%23111%3B%0A%20%20overflow%2Dx%3A%20inherit%3B%0A%7D%0A%0A%2F%2A%20markdown%20v1%20%2A%2F%0Apre%20code%5Bclass%5D%20%7B%0A%20%20background%2Dcolor%3A%20inherit%3B%0A%7D%0A%0A%2F%2A%20markdown%20v2%20%2A%2F%0Apre%5Bclass%5D%20code%20%7B%0A%20%20background%2Dcolor%3A%20inherit%3B%0A%7D%0A%0A%2F%2A%20formatting%20of%20inline%20code%20%2A%2F%0Acode%20%7B%20%0A%20%20background%2Dcolor%3A%20transparent%3B%0A%20%20color%3A%20%2387b13f%3B%0A%20%20font%2Dsize%3A%2092%25%3B%0A%7D%0A%0A%2F%2A%20formatting%20of%20tables%20%2A%2F%0A%0Atable%2C%20td%2C%20th%20%7B%0A%20%20border%3A%20none%3B%0A%20%20padding%3A%200%200%2E5em%3B%0A%7D%0A%0A%2F%2A%20alternating%20row%20colors%20%2A%2F%0Atbody%20tr%3Anth%2Dchild%28odd%29%20td%20%7B%0A%20%20background%2Dcolor%3A%20%23F8F8F8%3B%0A%7D%0A%0Ablockquote%20%7B%0A%20%20%20color%3A%23666666%3B%0A%20%20%20margin%3A0%3B%0A%20%20%20padding%2Dleft%3A%201em%3B%0A%20%20%20border%2Dleft%3A%200%2E5em%20%23EEE%20solid%3B%0A%7D%0A%0Ahr%20%7B%0A%20%20%20height%3A%200px%3B%0A%20%20%20border%2Dbottom%3A%20none%3B%0A%20%20%20border%2Dtop%2Dwidth%3A%20thin%3B%0A%20%20%20border%2Dtop%2Dstyle%3A%20dotted%3B%0A%20%20%20border%2Dtop%2Dcolor%3A%20%23999999%3B%0A%7D%0A%0Aspan%2Eheader%2Dsection%2Dnumber%20%7B%0A%20%20padding%2Dright%3A%201em%3B%0A%7D%0A%0Aspan%2Etoc%2Dsection%2Dnumber%3A%3Aafter%20%7B%0A%20%20%20%20content%3A%20%22%20%20%22%3B%0A%20%20%20%20white%2Dspace%3A%20pre%3B%0A%7D%0A%0A%40media%20print%20%7B%0A%20%20%20%2A%20%7B%0A%20%20%20%20%20%20background%3A%20transparent%20%21important%3B%0A%20%20%20%20%20%20color%3A%20black%20%21important%3B%0A%20%20%20%20%20%20filter%3Anone%20%21important%3B%0A%20%20%20%20%20%20%2Dms%2Dfilter%3A%20none%20%21important%3B%0A%20%20%20%7D%0A%0A%20%20%20body%20%7B%0A%20%20%20%20%20%20font%2Dsize%3A12pt%3B%0A%20%20%20%20%20%20max%2Dwidth%3A100%25%3B%0A%20%20%20%7D%0A%0A%20%20%20a%2C%20a%3Avisited%20%7B%0A%20%20%20%20%20%20text%2Ddecoration%3A%20underline%3B%0A%20%20%20%7D%0A%0A%20%20%20hr%20%7B%0A%20%20%20%20%20%20visibility%3A%20hidden%3B%0A%20%20%20%20%20%20page%2Dbreak%2Dbefore%3A%20always%3B%0A%20%20%20%7D%0A%0A%20%20%20pre%2C%20blockquote%20%7B%0A%20%20%20%20%20%20padding%2Dright%3A%201em%3B%0A%20%20%20%20%20%20page%2Dbreak%2Dinside%3A%20avoid%3B%0A%20%20%20%7D%0A%0A%20%20%20tr%2C%20img%20%7B%0A%20%20%20%20%20%20page%2Dbreak%2Dinside%3A%20avoid%3B%0A%20%20%20%7D%0A%0A%20%20%20img%20%7B%0A%20%20%20%20%20%20max%2Dwidth%3A%20100%25%20%21important%3B%0A%20%20%20%7D%0A%0A%20%20%20%40page%20%3Aleft%20%7B%0A%20%20%20%20%20%20margin%3A%2015mm%2020mm%2015mm%2010mm%3B%0A%20%20%20%7D%0A%0A%20%20%20%40page%20%3Aright%20%7B%0A%20%20%20%20%20%20margin%3A%2015mm%2010mm%2015mm%2020mm%3B%0A%20%20%20%7D%0A%0A%20%20%20p%2C%20h2%2C%20h3%20%7B%0A%20%20%20%20%20%20orphans%3A%203%3B%20widows%3A%203%3B%0A%20%20%20%7D%0A%0A%20%20%20h2%2C%20h3%20%7B%0A%20%20%20%20%20%20page%2Dbreak%2Dafter%3A%20avoid%3B%0A%20%20%20%7D%0A%7D%0A" rel="stylesheet" type="text/css" />

<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  var links = document.links;  
  for (var i = 0, linksLength = links.length; i < linksLength; i++)
    if(links[i].hostname != window.location.hostname)
      links[i].target = '_blank';
});
</script>

</head>

<body>


<div id="header">
<h1 class="title">Evaluate results</h1>
<h4 class="author"><em>L Collado-Torres</em></h4>
<h4 class="date"><em>21 February 2016</em></h4>
</div>

<h1>Contents</h1>
<div id="TOC">
<ul>
<li><a href="#setup"><span class="toc-section-number">1</span> Setup</a><ul>
<li><a href="#load-data"><span class="toc-section-number">1.1</span> Load data</a></li>
<li><a href="#evaluation-functions"><span class="toc-section-number">1.2</span> Evaluation functions</a></li>
<li><a href="#reference-sets"><span class="toc-section-number">1.3</span> Reference sets</a></li>
</ul></li>
<li><a href="#evaluation"><span class="toc-section-number">2</span> Evaluation</a><ul>
<li><a href="#evaluate-results"><span class="toc-section-number">2.1</span> Evaluate results</a></li>
<li><a href="#summaries"><span class="toc-section-number">2.2</span> Summaries</a></li>
<li><a href="#details"><span class="toc-section-number">2.3</span> Details</a></li>
<li><a href="#specific-case"><span class="toc-section-number">2.4</span> Specific case</a></li>
</ul></li>
<li><a href="#reproducibility"><span class="toc-section-number">3</span> Reproducibility</a></li>
</ul>
</div>

<p>Here we evaluate the performance of the different pipelines we used to analyze the simulated data. Evaluating the performance is tricky and depends on how you define the reference set. Here we show the results by using all exons and exons that overlap only 1 transcript. The second set removes ambiguity in determining the truth and is thus the most useful. Other possible reference sets can be considered and you can find the results in previous versions of this document. We removed them from the final version to avoid confusion.</p>
<div id="setup" class="section level1">
<h1><span class="header-section-number">1</span> Setup</h1>
<p>First, we load the required libraries.</p>
<pre class="r"><code>## Load required libraries
library('GenomicRanges')
library('TxDb.Hsapiens.UCSC.hg19.knownGene')
library('knitr')
library('devtools')</code></pre>
<div id="load-data" class="section level2">
<h2><span class="header-section-number">1.1</span> Load data</h2>
<p>Next we load the required data from the simulation setup as well as the statistical results from the different pipelines. We then format the results in a way that we can easily access them later on.</p>
<pre class="r"><code>## Load simulation information
load(file.path('..', 'generateReads', 'simulation_info.Rdata'))

## Load transcripts used
txdb &lt;- keepSeqlevels(TxDb.Hsapiens.UCSC.hg19.knownGene, 'chr17')
txinfo &lt;- select(txdb, keys = chosen$ucsckg_id, columns = columns(txdb), keytype = 'TXNAME')</code></pre>
<pre><code>## 'select()' returned 1:many mapping between keys and columns</code></pre>
<pre class="r"><code>## Build GRangesList with exons grouped by transcript
tx &lt;- split(GRanges(seqnames = txinfo$EXONCHROM, IRanges(start = txinfo$EXONSTART, end = txinfo$EXONEND), strand = txinfo$EXONSTRAND), txinfo$TXNAME)
tx &lt;- tx[match(chosen$ucsckg_id, names(tx))]


if(!identical(nrow(chosen), length(tx))) {
    message(paste(Sys.time(), 'Subsetting list of transcripts to those that were used when simulating the reads'))
    tx &lt;- tx[names(tx) %in% chosen$gene_id]
    stopifnot(unique(sum(width(tx)) - chosen$width) == 0)
}

## Load stat results
inputs &lt;- c('featureCounts', 'railMatrix', 'regionMatrix', 'ballgown')
reps &lt;- 1:3
softwares &lt;- c('DESeq2', 'edgeR')
statuses &lt;- c('complete', 'incomplete')

## Function for loading results from deseq2-edger folder
loadDE &lt;- function(software, file) {
    load(file.path('..', 'deseq2-edger', paste0('stats-', file, '-', software, '.Rdata')))
    if(software == 'DESeq2') {
        DEres &lt;- deseq
    } else {
        DEres &lt;- edger
    }
    return(DEres)
}

## Function for loading ballgown results
loadBG &lt;- function(level, file) {
    load(file.path('..', 'ballgown', paste0('bgres-', level, '-', file, '.Rdata')))
    return(bgres)
}

## Actually load the stats results
stats &lt;- lapply(inputs, function(input) {
    inputres &lt;- lapply(reps, function(rep) {
        if(input %in% c('featureCounts', 'ballgown')) {
            statusres &lt;- lapply(statuses, function(status) {
                if(input == 'ballgown') {
                    res &lt;- lapply(c('trans', 'exon'), loadBG, file = 
                        paste0('R', rep, ifelse(status == 'complete', '-comp', '-inc')))
                    names(res) &lt;- c('trans', 'exon')
                } else if (input == 'featureCounts') {
                    res &lt;- lapply(softwares, loadDE, file = paste0('featureCounts-R', rep, 
                        ifelse(status == 'complete', '-comp', '-inc')))
                    names(res) &lt;- softwares
                }
                return(res)      
            })
            names(statusres) &lt;- statuses
            return(statusres)
        } else {
            softres &lt;- lapply(softwares, loadDE, file = paste0(input, '-R', rep))
            names(softres) &lt;- softwares
            return(softres)
        }
    })
    names(inputres) &lt;- reps
    return(inputres)
})
names(stats) &lt;- inputs

## Group the stats results into a list without nested levels
stats_GR &lt;- lapply(inputs, function(input) {
    if(input %in% c('featureCounts', 'ballgown')) {
        res &lt;- lapply(stats[[input]], unlist)
    } else {
        res &lt;- unlist(stats[[input]])
    }
    res &lt;- unlist(res)
    return(res)
})
names(stats_GR) &lt;- inputs
stats_GR &lt;- unlist(stats_GR)</code></pre>
</div>
<div id="evaluation-functions" class="section level2">
<h2><span class="header-section-number">1.2</span> Evaluation functions</h2>
<p>The following code defines the functions we will use to evaluate the statistical results.</p>
<pre class="r"><code>## count_comp compares the information at hand versus the reference set
count_comp &lt;- function(info, rep = 1, type = 'padj', reference, cut = 0.05) {
    if(type == 'padj') {
        idx &lt;- mcols(info)$padj &lt; cut
    } else if (type == 'qval') {
        idx &lt;- mcols(info)$qval &lt; cut
    }
    
    idx[is.na(idx)] &lt;- FALSE

    ## Overlaps at least 1 DE 'reference' unit
    addmargins(table('DE truth' = mcols(reference)[[paste0('DEr', rep)]],
        'Overlaps DE' = countOverlaps(reference, info[idx]) &gt; 0))
}


## Functions for evaluating empirical power, FPR and FDR, plus summarizing 
## the info
emp_power &lt;- function(m, digits = 1) {
    round(m[2, 2] / m[2, 3] * 100, digits)
}
emp_fpr &lt;- function(m, digits = 1) {
    round(m[1, 2] / m[1, 3] * 100, digits)
}
emp_fdr &lt;- function(m, digits = 1) {
    round(m[1, 2] / m[3, 2] * 100, digits)
}

## Detailed table with the results for each replicate of the simulation
emp &lt;- function(tables) {
    empirical &lt;- data.frame(
        power = sapply(tables, emp_power),
        FPR = sapply(tables, emp_fpr),
        FDR = sapply(tables, emp_fdr)
    )
    empirical$replicate &lt;- as.integer(sapply(strsplit(names(tables), '\\.'), '[[', 2))
    empirical$AnnotationComplete &lt;- NA
    empirical$AnnotationComplete[grepl('\\.complete', names(tables))] &lt;- TRUE
    empirical$AnnotationComplete[grepl('\\.incomplete', names(tables))] &lt;- FALSE
    empirical$Aligner &lt;- 'HISAT'
    empirical$Aligner[grepl('rail', names(tables))] &lt;- 'Rail-RNA'
    empirical$SummaryMethod &lt;- 'StringTie'
    empirical$SummaryMethod[grepl('Matrix', names(tables))] &lt;- 'derfinder'
    empirical$SummaryMethod[grepl('featureCounts', names(tables))] &lt;- 'featureCounts'
    empirical$StatMethod &lt;- 'DESeq2'
    empirical$StatMethod[grepl('edgeR', names(tables))] &lt;- 'edgeR'
    empirical$StatMethod[grepl('ballgown.*exon', names(tables))] &lt;- 'ballgown-exon'
    empirical$StatMethod[grepl('ballgown.*trans', names(tables))] &lt;- 'ballgown-trans'
    rownames(empirical) &lt;- NULL
    return(empirical)
}

## This function takes the result from emp() and summarizes it by showing
## the minimum and maximum value per replicate
emp_sum_paste &lt;- function(x) { paste0('(', x[1], '-', x[2], ')') }
emp_sum &lt;- function(empinfo) {
    empinfo$situation &lt;- paste(empinfo$AnnotationComplete, empinfo$Aligner, 
        empinfo$SummaryMethod, empinfo$StatMethod, sep=';')
    empsum &lt;- data.frame(
        'Power' = sapply(tapply(empinfo$power, empinfo$situation, range),
            emp_sum_paste),
        'FPR' = sapply(tapply(empinfo$FPR, empinfo$situation, range),
            emp_sum_paste),
        'FDR' = sapply(tapply(empinfo$FDR, empinfo$situation, range),
            emp_sum_paste)
    )
    empsum$AnnotationComplete &lt;- as.logical(sapply(strsplit(rownames(empsum), ';'), '[[', 1))
    empsum$Aligner &lt;- sapply(strsplit(rownames(empsum), ';'), '[[', 2)
    empsum$SummaryMethod &lt;- sapply(strsplit(rownames(empsum), ';'), '[[', 3)
    empsum$StatMethod &lt;- sapply(strsplit(rownames(empsum), ';'), '[[', 4)
    empsum &lt;- empsum[c(1, 9, 2, 10, 3, 11, 4, 12, 5, 7, 6, 8), ]
    rownames(empsum) &lt;- NULL
    empsum
    
}

## index_comp is similar to count_comp but it returns the actual indices for
## the false positives, true positives, etc
index_comp &lt;- function(info, rep = 1, type = 'padj', reference, cut = 0.05) {
    if(type == 'padj') {
        idx &lt;- mcols(info)$padj &lt; cut
    } else if (type == 'qval') {
        idx &lt;- mcols(info)$qval &lt; cut
    }
    
    idx[is.na(idx)] &lt;- FALSE

    ## Overlaps at least 1 DE 'reference' unit
    ov &lt;- countOverlaps(reference, info[idx]) &gt; 0
    TP &lt;- mcols(reference)[[paste0('DEr', rep)]] &amp; ov
    TN &lt;- !mcols(reference)[[paste0('DEr', rep)]] &amp; !ov
    FP &lt;- !mcols(reference)[[paste0('DEr', rep)]] &amp; ov
    FN &lt;- mcols(reference)[[paste0('DEr', rep)]] &amp; !ov
    return(list(TruePositive = TP, TrueNegative = TN, FalsePositive = FP, FalseNegative = FN))
}

## case_result uses the information from index_comp to make a nice summary table
case_result &lt;- function(idx, r, reference) { 
    res &lt;- lapply(names(idx), function(i) {
        cases &lt;- mcols(reference)[[paste0('case', r)]][idx[[i]]]
        if(length(cases) == 0) return(NULL)
        data.frame(case = cases, result = i, stringsAsFactors = FALSE)
    })
    res &lt;- do.call(rbind, res)
    addmargins(table(&quot;Case&quot; = res$case, &quot;Result&quot; = res$result))
}

## Chose the most frequent case, used for choosing the case in the exons 
## reference sets.exon
mostFreq &lt;- function(cases) {
    if(length(cases) == 1) return(cases)
    names(sort(table(cases), decreasing = TRUE))[1]
}

## Process exons set to create a reference set
processExons &lt;- function(exons_set) {
    ov &lt;- findOverlaps(exons_set, tx)
    ## Find DE status per exon
    sHits &lt;- subjectHits(ov)
    qHits &lt;- queryHits(ov)
    exons_set$DEr1 &lt;- as.vector(tapply(sHits, qHits, function(y) {
        ## Note that we could change this from any() to all() and that
        ## would impact change whether the exon is labeled as DE or not.
        any(chosen$rep1[y] != 'normal')
    }))
    exons_set$DEr2 &lt;- as.vector(tapply(sHits, qHits, function(y) {
        any(chosen$rep2[y] != 'normal')
    }))
    exons_set$DEr3 &lt;- as.vector(tapply(sHits, qHits, function(y) {
        any(chosen$rep3[y] != 'normal')
    }))

    exons_set$cases1 &lt;- as.vector(tapply(sHits, qHits, function(y) {
        chosen$case1[y]
    }))
    exons_set$cases2 &lt;- as.vector(tapply(sHits, qHits, function(y) {
        chosen$case2[y]
    }))
    exons_set$cases3 &lt;- as.vector(tapply(sHits, qHits, function(y) {
        chosen$case3[y]
    }))
    exons_set$case1 &lt;- unname(sapply(exons_set$cases1, mostFreq))
    exons_set$case2 &lt;- unname(sapply(exons_set$cases2, mostFreq))
    exons_set$case3 &lt;- unname(sapply(exons_set$cases3, mostFreq))
    return(exons_set)
}</code></pre>
</div>
<div id="reference-sets" class="section level2">
<h2><span class="header-section-number">1.3</span> Reference sets</h2>
<p>We can now proceed to defining our reference sets. They are:</p>
<ol style="list-style-type: decimal">
<li>Exon level with DE status given by whether the exon overlaps any transcript with low or high expression. Note that if two transcripts share the same exon, that exon will appear twice in this reference set.</li>
<li>Exons that overlap only 1 transcript with DE status given by the transcript they overlap. These exons are not ambiguous in the truth assignment and is thus the most reliable reference set.</li>
</ol>
<pre class="r"><code>## Transcript level info
#trans &lt;- tx
#mcols(trans)$DEr1 &lt;- chosen$rep1 != 'normal'
#mcols(trans)$DEr2 &lt;- chosen$rep2 != 'normal'
#mcols(trans)$DEr3 &lt;- chosen$rep3 != 'normal'
#mcols(trans)$case1 &lt;- chosen$case1
#mcols(trans)$case2 &lt;- chosen$case2
#mcols(trans)$case3 &lt;- chosen$case3

## Transcript level information with DE status given by the gene level
#trans_case &lt;- trans
#mcols(trans_case)$DEr1 &lt;- chosen$case1 %in% c('allDE', 'singleDE', 'someDE')
#mcols(trans_case)$DEr2 &lt;- chosen$case2 %in% c('allDE', 'singleDE', 'someDE')
#mcols(trans_case)$DEr3 &lt;- chosen$case3 %in% c('allDE', 'singleDE', 'someDE')


## Create exon level reference
exons &lt;- processExons(unlist(tx))

## Create disjoint exon level reference
#exons_disjoin &lt;- processExons(disjoin(unlist(tx)))

## Exons overlapping only 1 transcript
exons_one &lt;- exons[countOverlaps(exons, tx) == 1]

## Disjoint exons overlapping only 1 transcript
#exons_disjoin_one &lt;- exons_disjoin[countOverlaps(exons_disjoin, trans) == 1]


## Explore the number and percent of DE cases
percDE &lt;- function(reference) {
    res &lt;- lapply(1:3, function(r) {
        round(table(mcols(reference)[[paste0('DEr', r)]]) / length(reference) * 100, 2)
    })
    names(res) &lt;- paste0('rep', 1:3)
    unlist(res)
}

## First at the transcript level, which matches exactly the 2/6 of transcripts 
## set to be DE (1/6 low, 1/6 high)
#percDE(trans)

## Next at the transcript level with DE status given by the gene level
## The percent DE is highly increased given that genes with more than 1
## transcript are likely to have at least 1 DE.
#percDE(trans_case)

## At the exon level, the percent with the truth set as DE is similar than a
## the transcript (DE by gene) level and is highly increased.
percDE(exons)</code></pre>
<pre><code>## rep1.FALSE  rep1.TRUE rep2.FALSE  rep2.TRUE rep3.FALSE  rep3.TRUE 
##      24.34      75.66      26.20      73.80      24.56      75.44</code></pre>
<pre class="r"><code>## Exons overlapping only 1 transcript. The percent is close to the one used in
## the simulation setup. That is: 2/6 of transcripts set to be DE (1/6 low, 1/6 
## high)
percDE(exons_one)</code></pre>
<pre><code>## rep1.FALSE  rep1.TRUE rep2.FALSE  rep2.TRUE rep3.FALSE  rep3.TRUE 
##      59.64      40.36      68.69      31.31      68.12      31.88</code></pre>
<pre class="r"><code>## For each reference set, see how many other units of that set each piece overlaps
table(countOverlaps(exons) - 1)</code></pre>
<pre><code>## 
##    0    1    2    3    4    5    6    7    8    9   10   11   12   13   14 
## 3868 5874 7205 5901 4788 3734 3039 1311 1075  405  424  161  107  162   63 
##   15   16   17   19   20   21   23   24   25   26   27   46   47 
##   70   71   36   64   21    1    1  303  439  189   22    1    3</code></pre>
<pre class="r"><code>table(countOverlaps(exons_one) - 1)</code></pre>
<pre><code>## 
##    0 
## 3868</code></pre>
<p>Note how the overlap between the different units of the same reference set is considerable for the complete exon reference set.</p>
<p>Number of units per reference level:</p>
<ul>
<li>Exon level: 39338</li>
<li>Exon level subsetted to those that overlap only 1 transcript: 3868</li>
</ul>
</div>
</div>
<div id="evaluation" class="section level1">
<h1><span class="header-section-number">2</span> Evaluation</h1>
<div id="evaluate-results" class="section level2">
<h2><span class="header-section-number">2.1</span> Evaluate results</h2>
<p>The following code actually runs the evaluation functions for the different reference sets.</p>
<pre class="r"><code>## Get type and replicate info
types &lt;- ifelse(grepl('ballgown', names(stats_GR)), 'qval', 'padj')
replicates &lt;- as.integer(sapply(strsplit(names(stats_GR), '\\.'), '[[', 2))


## Evaluate at the exons level
## DE if any of the transcripts it overlaps is DE
tables_exons &lt;- mapply(count_comp, stats_GR, replicates, types,
    SIMPLIFY = FALSE, MoreArgs = list(reference = exons, cut = 0.05))
empirical_exons &lt;- emp(tables_exons)
index_exons &lt;- mapply(index_comp, stats_GR, replicates, types,
    SIMPLIFY = FALSE, MoreArgs = list(reference = exons, cut = 0.05))
case_result_exons &lt;- mapply(case_result, index_exons, replicates,
    SIMPLIFY = FALSE, MoreArgs = list(reference = exons))
    
## Evaluate at the exons level (overlapping only 1 transcript)
## DE if any of the transcripts it overlaps is DE
tables_exons_one &lt;- mapply(count_comp, stats_GR, replicates, types,
    SIMPLIFY = FALSE, MoreArgs = list(reference = exons_one, cut = 0.05))
empirical_exons_one &lt;- emp(tables_exons_one)
index_exons_one &lt;- mapply(index_comp, stats_GR, replicates, types,
    SIMPLIFY = FALSE, MoreArgs = list(reference = exons_one, cut = 0.05))
case_result_exons_one &lt;- mapply(case_result, index_exons_one, replicates,
    SIMPLIFY = FALSE, MoreArgs = list(reference = exons_one))</code></pre>
</div>
<div id="summaries" class="section level2">
<h2><span class="header-section-number">2.2</span> Summaries</h2>
<p>Below are summaries of showing the minimum and maximum empirical power, false positive rate (FPR) and false discovery rate (FDR) per replicate for each of the analysis pipelines by the different reference sets.</p>
<pre class="r"><code>## Exon level
kable(emp_sum(empirical_exons), format = 'html')</code></pre>
<table>
<thead>
<tr>
<th style="text-align:left;">
Power
</th>
<th style="text-align:left;">
FPR
</th>
<th style="text-align:left;">
FDR
</th>
<th style="text-align:left;">
AnnotationComplete
</th>
<th style="text-align:left;">
Aligner
</th>
<th style="text-align:left;">
SummaryMethod
</th>
<th style="text-align:left;">
StatMethod
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
(72.2-74)
</td>
<td style="text-align:left;">
(5.2-6.7)
</td>
<td style="text-align:left;">
(2.3-2.8)
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
featureCounts
</td>
<td style="text-align:left;">
DESeq2
</td>
</tr>
<tr>
<td style="text-align:left;">
(73.4-75)
</td>
<td style="text-align:left;">
(5.6-6.6)
</td>
<td style="text-align:left;">
(2.6-2.9)
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
featureCounts
</td>
<td style="text-align:left;">
DESeq2
</td>
</tr>
<tr>
<td style="text-align:left;">
(77.4-79.6)
</td>
<td style="text-align:left;">
(10.9-12.6)
</td>
<td style="text-align:left;">
(4.4-5)
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
featureCounts
</td>
<td style="text-align:left;">
edgeR
</td>
</tr>
<tr>
<td style="text-align:left;">
(78.4-80.6)
</td>
<td style="text-align:left;">
(12.2-13.1)
</td>
<td style="text-align:left;">
(5-5.1)
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
featureCounts
</td>
<td style="text-align:left;">
edgeR
</td>
</tr>
<tr>
<td style="text-align:left;">
(70.9-72.9)
</td>
<td style="text-align:left;">
(3.1-4.1)
</td>
<td style="text-align:left;">
(1.3-1.8)
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
StringTie
</td>
<td style="text-align:left;">
ballgown-exon
</td>
</tr>
<tr>
<td style="text-align:left;">
(72-74.6)
</td>
<td style="text-align:left;">
(3.3-4.4)
</td>
<td style="text-align:left;">
(1.4-1.9)
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
StringTie
</td>
<td style="text-align:left;">
ballgown-exon
</td>
</tr>
<tr>
<td style="text-align:left;">
(47.1-51.9)
</td>
<td style="text-align:left;">
(1-1.8)
</td>
<td style="text-align:left;">
(0.7-1.2)
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
StringTie
</td>
<td style="text-align:left;">
ballgown-trans
</td>
</tr>
<tr>
<td style="text-align:left;">
(47.8-51.6)
</td>
<td style="text-align:left;">
(1.2-3.1)
</td>
<td style="text-align:left;">
(0.8-1.9)
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
StringTie
</td>
<td style="text-align:left;">
ballgown-trans
</td>
</tr>
<tr>
<td style="text-align:left;">
(71.6-73)
</td>
<td style="text-align:left;">
(8.3-10.1)
</td>
<td style="text-align:left;">
(3.6-4.2)
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
derfinder
</td>
<td style="text-align:left;">
DESeq2
</td>
</tr>
<tr>
<td style="text-align:left;">
(71.6-72.7)
</td>
<td style="text-align:left;">
(8.3-9.4)
</td>
<td style="text-align:left;">
(3.9-4)
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
Rail-RNA
</td>
<td style="text-align:left;">
derfinder
</td>
<td style="text-align:left;">
DESeq2
</td>
</tr>
<tr>
<td style="text-align:left;">
(72.6-73.4)
</td>
<td style="text-align:left;">
(9.9-10.8)
</td>
<td style="text-align:left;">
(4.3-4.9)
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
derfinder
</td>
<td style="text-align:left;">
edgeR
</td>
</tr>
<tr>
<td style="text-align:left;">
(72.4-73.2)
</td>
<td style="text-align:left;">
(10-10.6)
</td>
<td style="text-align:left;">
(4.4-4.7)
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
Rail-RNA
</td>
<td style="text-align:left;">
derfinder
</td>
<td style="text-align:left;">
edgeR
</td>
</tr>
</tbody>
</table>
<pre class="r"><code>## Exon level (overlapping only 1 transcript)
kable(emp_sum(empirical_exons_one), format = 'html')</code></pre>
<table>
<thead>
<tr>
<th style="text-align:left;">
Power
</th>
<th style="text-align:left;">
FPR
</th>
<th style="text-align:left;">
FDR
</th>
<th style="text-align:left;">
AnnotationComplete
</th>
<th style="text-align:left;">
Aligner
</th>
<th style="text-align:left;">
SummaryMethod
</th>
<th style="text-align:left;">
StatMethod
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
(70-78.3)
</td>
<td style="text-align:left;">
(2.1-3.6)
</td>
<td style="text-align:left;">
(5.6-9.8)
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
featureCounts
</td>
<td style="text-align:left;">
DESeq2
</td>
</tr>
<tr>
<td style="text-align:left;">
(95.1-95.9)
</td>
<td style="text-align:left;">
(2.9-4.8)
</td>
<td style="text-align:left;">
(6.3-9.7)
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
featureCounts
</td>
<td style="text-align:left;">
DESeq2
</td>
</tr>
<tr>
<td style="text-align:left;">
(71.2-79.2)
</td>
<td style="text-align:left;">
(5-8.1)
</td>
<td style="text-align:left;">
(12-19.5)
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
featureCounts
</td>
<td style="text-align:left;">
edgeR
</td>
</tr>
<tr>
<td style="text-align:left;">
(96.4-97.2)
</td>
<td style="text-align:left;">
(6.8-9.9)
</td>
<td style="text-align:left;">
(12.7-18.1)
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
featureCounts
</td>
<td style="text-align:left;">
edgeR
</td>
</tr>
<tr>
<td style="text-align:left;">
(68.4-77)
</td>
<td style="text-align:left;">
(2.8-3)
</td>
<td style="text-align:left;">
(5.5-8.3)
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
StringTie
</td>
<td style="text-align:left;">
ballgown-exon
</td>
</tr>
<tr>
<td style="text-align:left;">
(93.7-94.6)
</td>
<td style="text-align:left;">
(3.6-4)
</td>
<td style="text-align:left;">
(5.9-7.8)
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
StringTie
</td>
<td style="text-align:left;">
ballgown-exon
</td>
</tr>
<tr>
<td style="text-align:left;">
(53.2-60)
</td>
<td style="text-align:left;">
(0.6-2.2)
</td>
<td style="text-align:left;">
(1.4-8.1)
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
StringTie
</td>
<td style="text-align:left;">
ballgown-trans
</td>
</tr>
<tr>
<td style="text-align:left;">
(67.2-71.9)
</td>
<td style="text-align:left;">
(0.6-1.1)
</td>
<td style="text-align:left;">
(1.4-3.2)
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
StringTie
</td>
<td style="text-align:left;">
ballgown-trans
</td>
</tr>
<tr>
<td style="text-align:left;">
(94.3-95.5)
</td>
<td style="text-align:left;">
(6.8-9)
</td>
<td style="text-align:left;">
(12.2-16.1)
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
derfinder
</td>
<td style="text-align:left;">
DESeq2
</td>
</tr>
<tr>
<td style="text-align:left;">
(94.1-95.1)
</td>
<td style="text-align:left;">
(6.5-8.3)
</td>
<td style="text-align:left;">
(11.1-15.7)
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
Rail-RNA
</td>
<td style="text-align:left;">
derfinder
</td>
<td style="text-align:left;">
DESeq2
</td>
</tr>
<tr>
<td style="text-align:left;">
(95.6-96.1)
</td>
<td style="text-align:left;">
(8.4-10.5)
</td>
<td style="text-align:left;">
(13.9-18.9)
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
derfinder
</td>
<td style="text-align:left;">
edgeR
</td>
</tr>
<tr>
<td style="text-align:left;">
(94.7-95.9)
</td>
<td style="text-align:left;">
(8.3-10.2)
</td>
<td style="text-align:left;">
(12.8-18.5)
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
Rail-RNA
</td>
<td style="text-align:left;">
derfinder
</td>
<td style="text-align:left;">
edgeR
</td>
</tr>
</tbody>
</table>
<p>From these summaries we can see that in most cases the minimum and maximum values are nearly the same, with more variability for the <code>HISAT</code> -&gt; <code>StringTie</code> -&gt; <code>ballgown</code> pipeline results. Note that for this pipeline we conducted tests at the transcript level (StatMethod = <code>ballgown-trans</code>) and at the exon level (<code>ballgown-exon</code>). <code>edgeR</code> achieves slightly greater empirical power than <code>DESeq2</code> at the cost of a higher empirical FPR and FDR.</p>
<p>The sets where we consider exons overlapping only 1 transcript shows a clear difference in power for the methods that rely on annotation between the analyses that used the complete annotation and those that used the incomplete one. The <code>derfinder</code> analyses do have higher FPR and FDR than other pipelines, but achieve nearly the same power as the other methods when the annotation is complete.</p>
</div>
<div id="details" class="section level2">
<h2><span class="header-section-number">2.3</span> Details</h2>
<p>The following tables show the results for all the replicates.</p>
<pre class="r"><code>## Exon level
kable(empirical_exons, format = 'html', row.names = TRUE)</code></pre>
<table>
<thead>
<tr>
<th style="text-align:left;">
</th>
<th style="text-align:right;">
power
</th>
<th style="text-align:right;">
FPR
</th>
<th style="text-align:right;">
FDR
</th>
<th style="text-align:right;">
replicate
</th>
<th style="text-align:left;">
AnnotationComplete
</th>
<th style="text-align:left;">
Aligner
</th>
<th style="text-align:left;">
SummaryMethod
</th>
<th style="text-align:left;">
StatMethod
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
1
</td>
<td style="text-align:right;">
75.0
</td>
<td style="text-align:right;">
6.6
</td>
<td style="text-align:right;">
2.8
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
featureCounts
</td>
<td style="text-align:left;">
DESeq2
</td>
</tr>
<tr>
<td style="text-align:left;">
2
</td>
<td style="text-align:right;">
80.6
</td>
<td style="text-align:right;">
13.1
</td>
<td style="text-align:right;">
5.0
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
featureCounts
</td>
<td style="text-align:left;">
edgeR
</td>
</tr>
<tr>
<td style="text-align:left;">
3
</td>
<td style="text-align:right;">
74.0
</td>
<td style="text-align:right;">
6.7
</td>
<td style="text-align:right;">
2.8
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
featureCounts
</td>
<td style="text-align:left;">
DESeq2
</td>
</tr>
<tr>
<td style="text-align:left;">
4
</td>
<td style="text-align:right;">
79.6
</td>
<td style="text-align:right;">
12.6
</td>
<td style="text-align:right;">
4.8
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
featureCounts
</td>
<td style="text-align:left;">
edgeR
</td>
</tr>
<tr>
<td style="text-align:left;">
5
</td>
<td style="text-align:right;">
74.0
</td>
<td style="text-align:right;">
5.6
</td>
<td style="text-align:right;">
2.6
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
featureCounts
</td>
<td style="text-align:left;">
DESeq2
</td>
</tr>
<tr>
<td style="text-align:left;">
6
</td>
<td style="text-align:right;">
80.0
</td>
<td style="text-align:right;">
12.2
</td>
<td style="text-align:right;">
5.1
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
featureCounts
</td>
<td style="text-align:left;">
edgeR
</td>
</tr>
<tr>
<td style="text-align:left;">
7
</td>
<td style="text-align:right;">
72.9
</td>
<td style="text-align:right;">
5.2
</td>
<td style="text-align:right;">
2.5
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
featureCounts
</td>
<td style="text-align:left;">
DESeq2
</td>
</tr>
<tr>
<td style="text-align:left;">
8
</td>
<td style="text-align:right;">
78.9
</td>
<td style="text-align:right;">
11.7
</td>
<td style="text-align:right;">
5.0
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
featureCounts
</td>
<td style="text-align:left;">
edgeR
</td>
</tr>
<tr>
<td style="text-align:left;">
9
</td>
<td style="text-align:right;">
73.4
</td>
<td style="text-align:right;">
6.6
</td>
<td style="text-align:right;">
2.9
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
featureCounts
</td>
<td style="text-align:left;">
DESeq2
</td>
</tr>
<tr>
<td style="text-align:left;">
10
</td>
<td style="text-align:right;">
78.4
</td>
<td style="text-align:right;">
12.9
</td>
<td style="text-align:right;">
5.1
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
featureCounts
</td>
<td style="text-align:left;">
edgeR
</td>
</tr>
<tr>
<td style="text-align:left;">
11
</td>
<td style="text-align:right;">
72.2
</td>
<td style="text-align:right;">
5.2
</td>
<td style="text-align:right;">
2.3
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
featureCounts
</td>
<td style="text-align:left;">
DESeq2
</td>
</tr>
<tr>
<td style="text-align:left;">
12
</td>
<td style="text-align:right;">
77.4
</td>
<td style="text-align:right;">
10.9
</td>
<td style="text-align:right;">
4.4
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
featureCounts
</td>
<td style="text-align:left;">
edgeR
</td>
</tr>
<tr>
<td style="text-align:left;">
13
</td>
<td style="text-align:right;">
72.7
</td>
<td style="text-align:right;">
9.4
</td>
<td style="text-align:right;">
4.0
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
Rail-RNA
</td>
<td style="text-align:left;">
derfinder
</td>
<td style="text-align:left;">
DESeq2
</td>
</tr>
<tr>
<td style="text-align:left;">
14
</td>
<td style="text-align:right;">
73.2
</td>
<td style="text-align:right;">
10.6
</td>
<td style="text-align:right;">
4.5
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
Rail-RNA
</td>
<td style="text-align:left;">
derfinder
</td>
<td style="text-align:left;">
edgeR
</td>
</tr>
<tr>
<td style="text-align:left;">
15
</td>
<td style="text-align:right;">
71.6
</td>
<td style="text-align:right;">
8.3
</td>
<td style="text-align:right;">
4.0
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
Rail-RNA
</td>
<td style="text-align:left;">
derfinder
</td>
<td style="text-align:left;">
DESeq2
</td>
</tr>
<tr>
<td style="text-align:left;">
16
</td>
<td style="text-align:right;">
72.4
</td>
<td style="text-align:right;">
10.0
</td>
<td style="text-align:right;">
4.7
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
Rail-RNA
</td>
<td style="text-align:left;">
derfinder
</td>
<td style="text-align:left;">
edgeR
</td>
</tr>
<tr>
<td style="text-align:left;">
17
</td>
<td style="text-align:right;">
72.3
</td>
<td style="text-align:right;">
8.9
</td>
<td style="text-align:right;">
3.9
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
Rail-RNA
</td>
<td style="text-align:left;">
derfinder
</td>
<td style="text-align:left;">
DESeq2
</td>
</tr>
<tr>
<td style="text-align:left;">
18
</td>
<td style="text-align:right;">
73.0
</td>
<td style="text-align:right;">
10.3
</td>
<td style="text-align:right;">
4.4
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
Rail-RNA
</td>
<td style="text-align:left;">
derfinder
</td>
<td style="text-align:left;">
edgeR
</td>
</tr>
<tr>
<td style="text-align:left;">
19
</td>
<td style="text-align:right;">
73.0
</td>
<td style="text-align:right;">
10.1
</td>
<td style="text-align:right;">
4.2
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
derfinder
</td>
<td style="text-align:left;">
DESeq2
</td>
</tr>
<tr>
<td style="text-align:left;">
20
</td>
<td style="text-align:right;">
73.4
</td>
<td style="text-align:right;">
10.8
</td>
<td style="text-align:right;">
4.5
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
derfinder
</td>
<td style="text-align:left;">
edgeR
</td>
</tr>
<tr>
<td style="text-align:left;">
21
</td>
<td style="text-align:right;">
71.6
</td>
<td style="text-align:right;">
8.8
</td>
<td style="text-align:right;">
4.2
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
derfinder
</td>
<td style="text-align:left;">
DESeq2
</td>
</tr>
<tr>
<td style="text-align:left;">
22
</td>
<td style="text-align:right;">
72.6
</td>
<td style="text-align:right;">
10.6
</td>
<td style="text-align:right;">
4.9
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
derfinder
</td>
<td style="text-align:left;">
edgeR
</td>
</tr>
<tr>
<td style="text-align:left;">
23
</td>
<td style="text-align:right;">
72.3
</td>
<td style="text-align:right;">
8.3
</td>
<td style="text-align:right;">
3.6
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
derfinder
</td>
<td style="text-align:left;">
DESeq2
</td>
</tr>
<tr>
<td style="text-align:left;">
24
</td>
<td style="text-align:right;">
72.8
</td>
<td style="text-align:right;">
9.9
</td>
<td style="text-align:right;">
4.3
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
derfinder
</td>
<td style="text-align:left;">
edgeR
</td>
</tr>
<tr>
<td style="text-align:left;">
25
</td>
<td style="text-align:right;">
50.7
</td>
<td style="text-align:right;">
3.1
</td>
<td style="text-align:right;">
1.9
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
StringTie
</td>
<td style="text-align:left;">
ballgown-trans
</td>
</tr>
<tr>
<td style="text-align:left;">
26
</td>
<td style="text-align:right;">
72.0
</td>
<td style="text-align:right;">
4.4
</td>
<td style="text-align:right;">
1.9
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
StringTie
</td>
<td style="text-align:left;">
ballgown-exon
</td>
</tr>
<tr>
<td style="text-align:left;">
27
</td>
<td style="text-align:right;">
47.1
</td>
<td style="text-align:right;">
1.5
</td>
<td style="text-align:right;">
1.0
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
StringTie
</td>
<td style="text-align:left;">
ballgown-trans
</td>
</tr>
<tr>
<td style="text-align:left;">
28
</td>
<td style="text-align:right;">
70.9
</td>
<td style="text-align:right;">
4.1
</td>
<td style="text-align:right;">
1.8
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
StringTie
</td>
<td style="text-align:left;">
ballgown-exon
</td>
</tr>
<tr>
<td style="text-align:left;">
29
</td>
<td style="text-align:right;">
51.6
</td>
<td style="text-align:right;">
1.2
</td>
<td style="text-align:right;">
0.8
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
StringTie
</td>
<td style="text-align:left;">
ballgown-trans
</td>
</tr>
<tr>
<td style="text-align:left;">
30
</td>
<td style="text-align:right;">
74.1
</td>
<td style="text-align:right;">
3.7
</td>
<td style="text-align:right;">
1.7
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
StringTie
</td>
<td style="text-align:left;">
ballgown-exon
</td>
</tr>
<tr>
<td style="text-align:left;">
31
</td>
<td style="text-align:right;">
51.9
</td>
<td style="text-align:right;">
1.8
</td>
<td style="text-align:right;">
1.2
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
StringTie
</td>
<td style="text-align:left;">
ballgown-trans
</td>
</tr>
<tr>
<td style="text-align:left;">
32
</td>
<td style="text-align:right;">
72.9
</td>
<td style="text-align:right;">
3.5
</td>
<td style="text-align:right;">
1.7
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
StringTie
</td>
<td style="text-align:left;">
ballgown-exon
</td>
</tr>
<tr>
<td style="text-align:left;">
33
</td>
<td style="text-align:right;">
47.8
</td>
<td style="text-align:right;">
1.5
</td>
<td style="text-align:right;">
1.0
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
StringTie
</td>
<td style="text-align:left;">
ballgown-trans
</td>
</tr>
<tr>
<td style="text-align:left;">
34
</td>
<td style="text-align:right;">
74.6
</td>
<td style="text-align:right;">
3.3
</td>
<td style="text-align:right;">
1.4
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
StringTie
</td>
<td style="text-align:left;">
ballgown-exon
</td>
</tr>
<tr>
<td style="text-align:left;">
35
</td>
<td style="text-align:right;">
50.6
</td>
<td style="text-align:right;">
1.0
</td>
<td style="text-align:right;">
0.7
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
StringTie
</td>
<td style="text-align:left;">
ballgown-trans
</td>
</tr>
<tr>
<td style="text-align:left;">
36
</td>
<td style="text-align:right;">
72.9
</td>
<td style="text-align:right;">
3.1
</td>
<td style="text-align:right;">
1.3
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
StringTie
</td>
<td style="text-align:left;">
ballgown-exon
</td>
</tr>
</tbody>
</table>
<pre class="r"><code>## Exon level (overlapping only 1 transcript)
kable(empirical_exons_one, format = 'html', row.names = TRUE)</code></pre>
<table>
<thead>
<tr>
<th style="text-align:left;">
</th>
<th style="text-align:right;">
power
</th>
<th style="text-align:right;">
FPR
</th>
<th style="text-align:right;">
FDR
</th>
<th style="text-align:right;">
replicate
</th>
<th style="text-align:left;">
AnnotationComplete
</th>
<th style="text-align:left;">
Aligner
</th>
<th style="text-align:left;">
SummaryMethod
</th>
<th style="text-align:left;">
StatMethod
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
1
</td>
<td style="text-align:right;">
95.9
</td>
<td style="text-align:right;">
4.5
</td>
<td style="text-align:right;">
6.5
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
featureCounts
</td>
<td style="text-align:left;">
DESeq2
</td>
</tr>
<tr>
<td style="text-align:left;">
2
</td>
<td style="text-align:right;">
97.2
</td>
<td style="text-align:right;">
9.5
</td>
<td style="text-align:right;">
12.7
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
featureCounts
</td>
<td style="text-align:left;">
edgeR
</td>
</tr>
<tr>
<td style="text-align:left;">
3
</td>
<td style="text-align:right;">
78.3
</td>
<td style="text-align:right;">
3.5
</td>
<td style="text-align:right;">
6.2
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
featureCounts
</td>
<td style="text-align:left;">
DESeq2
</td>
</tr>
<tr>
<td style="text-align:left;">
4
</td>
<td style="text-align:right;">
79.2
</td>
<td style="text-align:right;">
7.3
</td>
<td style="text-align:right;">
12.0
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
featureCounts
</td>
<td style="text-align:left;">
edgeR
</td>
</tr>
<tr>
<td style="text-align:left;">
5
</td>
<td style="text-align:right;">
95.1
</td>
<td style="text-align:right;">
2.9
</td>
<td style="text-align:right;">
6.3
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
featureCounts
</td>
<td style="text-align:left;">
DESeq2
</td>
</tr>
<tr>
<td style="text-align:left;">
6
</td>
<td style="text-align:right;">
96.4
</td>
<td style="text-align:right;">
6.8
</td>
<td style="text-align:right;">
13.5
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
featureCounts
</td>
<td style="text-align:left;">
edgeR
</td>
</tr>
<tr>
<td style="text-align:left;">
7
</td>
<td style="text-align:right;">
77.5
</td>
<td style="text-align:right;">
2.1
</td>
<td style="text-align:right;">
5.6
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
featureCounts
</td>
<td style="text-align:left;">
DESeq2
</td>
</tr>
<tr>
<td style="text-align:left;">
8
</td>
<td style="text-align:right;">
78.4
</td>
<td style="text-align:right;">
5.0
</td>
<td style="text-align:right;">
12.3
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
featureCounts
</td>
<td style="text-align:left;">
edgeR
</td>
</tr>
<tr>
<td style="text-align:left;">
9
</td>
<td style="text-align:right;">
95.4
</td>
<td style="text-align:right;">
4.8
</td>
<td style="text-align:right;">
9.7
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
featureCounts
</td>
<td style="text-align:left;">
DESeq2
</td>
</tr>
<tr>
<td style="text-align:left;">
10
</td>
<td style="text-align:right;">
96.4
</td>
<td style="text-align:right;">
9.9
</td>
<td style="text-align:right;">
18.1
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
featureCounts
</td>
<td style="text-align:left;">
edgeR
</td>
</tr>
<tr>
<td style="text-align:left;">
11
</td>
<td style="text-align:right;">
70.0
</td>
<td style="text-align:right;">
3.6
</td>
<td style="text-align:right;">
9.8
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
featureCounts
</td>
<td style="text-align:left;">
DESeq2
</td>
</tr>
<tr>
<td style="text-align:left;">
12
</td>
<td style="text-align:right;">
71.2
</td>
<td style="text-align:right;">
8.1
</td>
<td style="text-align:right;">
19.5
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
featureCounts
</td>
<td style="text-align:left;">
edgeR
</td>
</tr>
<tr>
<td style="text-align:left;">
13
</td>
<td style="text-align:right;">
95.0
</td>
<td style="text-align:right;">
8.0
</td>
<td style="text-align:right;">
11.1
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
Rail-RNA
</td>
<td style="text-align:left;">
derfinder
</td>
<td style="text-align:left;">
DESeq2
</td>
</tr>
<tr>
<td style="text-align:left;">
14
</td>
<td style="text-align:right;">
95.3
</td>
<td style="text-align:right;">
9.4
</td>
<td style="text-align:right;">
12.8
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
Rail-RNA
</td>
<td style="text-align:left;">
derfinder
</td>
<td style="text-align:left;">
edgeR
</td>
</tr>
<tr>
<td style="text-align:left;">
15
</td>
<td style="text-align:right;">
94.1
</td>
<td style="text-align:right;">
6.5
</td>
<td style="text-align:right;">
13.1
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
Rail-RNA
</td>
<td style="text-align:left;">
derfinder
</td>
<td style="text-align:left;">
DESeq2
</td>
</tr>
<tr>
<td style="text-align:left;">
16
</td>
<td style="text-align:right;">
94.7
</td>
<td style="text-align:right;">
8.3
</td>
<td style="text-align:right;">
16.1
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
Rail-RNA
</td>
<td style="text-align:left;">
derfinder
</td>
<td style="text-align:left;">
edgeR
</td>
</tr>
<tr>
<td style="text-align:left;">
17
</td>
<td style="text-align:right;">
95.1
</td>
<td style="text-align:right;">
8.3
</td>
<td style="text-align:right;">
15.7
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
Rail-RNA
</td>
<td style="text-align:left;">
derfinder
</td>
<td style="text-align:left;">
DESeq2
</td>
</tr>
<tr>
<td style="text-align:left;">
18
</td>
<td style="text-align:right;">
95.9
</td>
<td style="text-align:right;">
10.2
</td>
<td style="text-align:right;">
18.5
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
Rail-RNA
</td>
<td style="text-align:left;">
derfinder
</td>
<td style="text-align:left;">
edgeR
</td>
</tr>
<tr>
<td style="text-align:left;">
19
</td>
<td style="text-align:right;">
95.5
</td>
<td style="text-align:right;">
9.0
</td>
<td style="text-align:right;">
12.2
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
derfinder
</td>
<td style="text-align:left;">
DESeq2
</td>
</tr>
<tr>
<td style="text-align:left;">
20
</td>
<td style="text-align:right;">
95.8
</td>
<td style="text-align:right;">
10.5
</td>
<td style="text-align:right;">
13.9
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
derfinder
</td>
<td style="text-align:left;">
edgeR
</td>
</tr>
<tr>
<td style="text-align:left;">
21
</td>
<td style="text-align:right;">
94.3
</td>
<td style="text-align:right;">
6.8
</td>
<td style="text-align:right;">
13.7
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
derfinder
</td>
<td style="text-align:left;">
DESeq2
</td>
</tr>
<tr>
<td style="text-align:left;">
22
</td>
<td style="text-align:right;">
95.6
</td>
<td style="text-align:right;">
8.4
</td>
<td style="text-align:right;">
16.1
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
derfinder
</td>
<td style="text-align:left;">
edgeR
</td>
</tr>
<tr>
<td style="text-align:left;">
23
</td>
<td style="text-align:right;">
95.3
</td>
<td style="text-align:right;">
8.5
</td>
<td style="text-align:right;">
16.1
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
derfinder
</td>
<td style="text-align:left;">
DESeq2
</td>
</tr>
<tr>
<td style="text-align:left;">
24
</td>
<td style="text-align:right;">
96.1
</td>
<td style="text-align:right;">
10.5
</td>
<td style="text-align:right;">
18.9
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
derfinder
</td>
<td style="text-align:left;">
edgeR
</td>
</tr>
<tr>
<td style="text-align:left;">
25
</td>
<td style="text-align:right;">
70.9
</td>
<td style="text-align:right;">
0.7
</td>
<td style="text-align:right;">
1.4
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
StringTie
</td>
<td style="text-align:left;">
ballgown-trans
</td>
</tr>
<tr>
<td style="text-align:left;">
26
</td>
<td style="text-align:right;">
94.6
</td>
<td style="text-align:right;">
4.0
</td>
<td style="text-align:right;">
5.9
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
StringTie
</td>
<td style="text-align:left;">
ballgown-exon
</td>
</tr>
<tr>
<td style="text-align:left;">
27
</td>
<td style="text-align:right;">
60.0
</td>
<td style="text-align:right;">
0.6
</td>
<td style="text-align:right;">
1.4
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
StringTie
</td>
<td style="text-align:left;">
ballgown-trans
</td>
</tr>
<tr>
<td style="text-align:left;">
28
</td>
<td style="text-align:right;">
77.0
</td>
<td style="text-align:right;">
3.0
</td>
<td style="text-align:right;">
5.5
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
StringTie
</td>
<td style="text-align:left;">
ballgown-exon
</td>
</tr>
<tr>
<td style="text-align:left;">
29
</td>
<td style="text-align:right;">
71.9
</td>
<td style="text-align:right;">
1.1
</td>
<td style="text-align:right;">
3.2
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
StringTie
</td>
<td style="text-align:left;">
ballgown-trans
</td>
</tr>
<tr>
<td style="text-align:left;">
30
</td>
<td style="text-align:right;">
94.3
</td>
<td style="text-align:right;">
3.6
</td>
<td style="text-align:right;">
7.8
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
StringTie
</td>
<td style="text-align:left;">
ballgown-exon
</td>
</tr>
<tr>
<td style="text-align:left;">
31
</td>
<td style="text-align:right;">
58.1
</td>
<td style="text-align:right;">
2.0
</td>
<td style="text-align:right;">
6.9
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
StringTie
</td>
<td style="text-align:left;">
ballgown-trans
</td>
</tr>
<tr>
<td style="text-align:left;">
32
</td>
<td style="text-align:right;">
77.0
</td>
<td style="text-align:right;">
2.8
</td>
<td style="text-align:right;">
7.4
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
StringTie
</td>
<td style="text-align:left;">
ballgown-exon
</td>
</tr>
<tr>
<td style="text-align:left;">
33
</td>
<td style="text-align:right;">
67.2
</td>
<td style="text-align:right;">
0.6
</td>
<td style="text-align:right;">
1.8
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
StringTie
</td>
<td style="text-align:left;">
ballgown-trans
</td>
</tr>
<tr>
<td style="text-align:left;">
34
</td>
<td style="text-align:right;">
93.7
</td>
<td style="text-align:right;">
3.6
</td>
<td style="text-align:right;">
7.7
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
StringTie
</td>
<td style="text-align:left;">
ballgown-exon
</td>
</tr>
<tr>
<td style="text-align:left;">
35
</td>
<td style="text-align:right;">
53.2
</td>
<td style="text-align:right;">
2.2
</td>
<td style="text-align:right;">
8.1
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
StringTie
</td>
<td style="text-align:left;">
ballgown-trans
</td>
</tr>
<tr>
<td style="text-align:left;">
36
</td>
<td style="text-align:right;">
68.4
</td>
<td style="text-align:right;">
2.9
</td>
<td style="text-align:right;">
8.3
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
HISAT
</td>
<td style="text-align:left;">
StringTie
</td>
<td style="text-align:left;">
ballgown-exon
</td>
</tr>
</tbody>
</table>
</div>
<div id="specific-case" class="section level2">
<h2><span class="header-section-number">2.4</span> Specific case</h2>
<p>In this section we focus on the specific case of using <code>HISAT</code> -&gt; <code>derfinder</code> -&gt; <code>edgeR</code> from the first replicate because it has the highest empirical FPR from all the <code>derfinder</code> analyses when using the exon reference set (row 20 in the first table).</p>
<p>When using the exon level reference set, only about a tenth [(37 + 642) / 6018 = 11.28%] of the <code>noneDE</code> cases are incorrectly labeled. The numbers decrease in the subset of exons that overlap only 1 transcript: [(0 + 37) / 448 = 8.26%].</p>
<pre class="r"><code>## Compare all references for one simulation scenario
case_result_exons[['regionMatrix.1.edgeR']]</code></pre>
<pre><code>##              Result
## Case          FalseNegative FalsePositive TrueNegative TruePositive   Sum
##   allDE                  81             0            0          595   676
##   noneDE                 37           642         5304           35  6018
##   singleDE               35             0            0          885   920
##   singleNotDE             0           149         1127            9  1285
##   someDE               7775           244         2110        20310 30439
##   Sum                  7928          1035         8541        21834 39338</code></pre>
<pre class="r"><code>case_result_exons_one[['regionMatrix.1.edgeR']]</code></pre>
<pre><code>##              Result
## Case          FalseNegative FalsePositive TrueNegative TruePositive  Sum
##   allDE                   2             0            0          137  139
##   noneDE                  0            37          411            0  448
##   singleDE               35             0            0          870  905
##   singleNotDE             0           145         1117            0 1262
##   someDE                 28            60          537          489 1114
##   Sum                    65           242         2065         1496 3868</code></pre>
<p>Next we show the same tables but in percent of the total reference set units.</p>
<pre class="r"><code>## Now in percent
round(case_result_exons[['regionMatrix.1.edgeR']] /
    case_result_exons[['regionMatrix.1.edgeR']][6, 5] * 100, 2)</code></pre>
<pre><code>##              Result
## Case          FalseNegative FalsePositive TrueNegative TruePositive    Sum
##   allDE                0.21          0.00         0.00         1.51   1.72
##   noneDE               0.09          1.63        13.48         0.09  15.30
##   singleDE             0.09          0.00         0.00         2.25   2.34
##   singleNotDE          0.00          0.38         2.86         0.02   3.27
##   someDE              19.76          0.62         5.36        51.63  77.38
##   Sum                 20.15          2.63        21.71        55.50 100.00</code></pre>
<pre class="r"><code>round(case_result_exons_one[['regionMatrix.1.edgeR']] /
    case_result_exons_one[['regionMatrix.1.edgeR']][6, 5] * 100, 2)</code></pre>
<pre><code>##              Result
## Case          FalseNegative FalsePositive TrueNegative TruePositive    Sum
##   allDE                0.05          0.00         0.00         3.54   3.59
##   noneDE               0.00          0.96        10.63         0.00  11.58
##   singleDE             0.90          0.00         0.00        22.49  23.40
##   singleNotDE          0.00          3.75        28.88         0.00  32.63
##   someDE               0.72          1.55        13.88        12.64  28.80
##   Sum                  1.68          6.26        53.39        38.68 100.00</code></pre>
<p>In both the exons reference set, most of the false negatives are from the <code>someDE</code> scenario. Those are ambiguous cases and it’s the main reason why the power is decreased when using this reference set. Note how the percent of total reference units from the <code>someDE</code> case that result in false negatives decreases from 19.76 to 0.72 between the exons set that does not take into account multiplicity and the subset of exons overlapping only 1 transcript.</p>
</div>
</div>
<div id="reproducibility" class="section level1">
<h1><span class="header-section-number">3</span> Reproducibility</h1>
<pre class="r"><code>## Reproducibility info
Sys.time()</code></pre>
<pre><code>## [1] &quot;2016-02-21 15:45:57 EST&quot;</code></pre>
<pre class="r"><code>proc.time()</code></pre>
<pre><code>##    user  system elapsed 
##  88.347   2.596 101.239</code></pre>
<pre class="r"><code>options(width = 120)
session_info()</code></pre>
<pre><code>## Session info -----------------------------------------------------------------------------------------------------------</code></pre>
<pre><code>##  setting  value                       
##  version  R version 3.2.2 (2015-08-14)
##  system   x86_64, darwin13.4.0        
##  ui       X11                         
##  language (EN)                        
##  collate  en_US.UTF-8                 
##  tz       America/New_York            
##  date     2016-02-21</code></pre>
<pre><code>## Packages ---------------------------------------------------------------------------------------------------------------</code></pre>
<pre><code>##  package                           * version  date       source        
##  AnnotationDbi                     * 1.32.3   2015-12-24 Bioconductor  
##  Biobase                           * 2.30.0   2015-10-14 Bioconductor  
##  BiocGenerics                      * 0.16.1   2015-11-06 Bioconductor  
##  BiocParallel                        1.4.3    2015-12-16 Bioconductor  
##  BiocStyle                         * 1.8.0    2015-10-14 Bioconductor  
##  biomaRt                             2.26.1   2015-11-23 Bioconductor  
##  Biostrings                          2.38.4   2016-02-09 Bioconductor  
##  bitops                              1.0-6    2013-08-17 CRAN (R 3.2.0)
##  colorout                          * 1.0-2    2015-05-19 local         
##  DBI                                 0.3.1    2014-09-24 CRAN (R 3.2.0)
##  devtools                          * 1.10.0   2016-01-23 CRAN (R 3.2.3)
##  digest                              0.6.9    2016-01-08 CRAN (R 3.2.3)
##  evaluate                            0.8      2015-09-18 CRAN (R 3.2.0)
##  formatR                             1.2.1    2015-09-18 CRAN (R 3.2.0)
##  futile.logger                       1.4.1    2015-04-20 CRAN (R 3.2.0)
##  futile.options                      1.0.0    2010-04-06 CRAN (R 3.2.0)
##  GenomeInfoDb                      * 1.6.3    2016-01-26 Bioconductor  
##  GenomicAlignments                   1.6.3    2016-01-06 Bioconductor  
##  GenomicFeatures                   * 1.22.13  2016-02-11 Bioconductor  
##  GenomicRanges                     * 1.22.4   2016-01-30 Bioconductor  
##  highr                               0.5.1    2015-09-18 CRAN (R 3.2.0)
##  htmltools                           0.3      2015-12-29 CRAN (R 3.2.3)
##  IRanges                           * 2.4.7    2016-02-09 Bioconductor  
##  knitr                             * 1.12.3   2016-01-22 CRAN (R 3.2.3)
##  lambda.r                            1.1.7    2015-03-20 CRAN (R 3.2.0)
##  magrittr                            1.5      2014-11-22 CRAN (R 3.2.0)
##  memoise                             1.0.0    2016-01-29 CRAN (R 3.2.3)
##  RCurl                               1.95-4.7 2015-06-30 CRAN (R 3.2.1)
##  rmarkdown                           0.9.2    2016-01-01 CRAN (R 3.2.3)
##  Rsamtools                           1.22.0   2015-10-14 Bioconductor  
##  RSQLite                             1.0.0    2014-10-25 CRAN (R 3.2.0)
##  rtracklayer                         1.30.2   2016-02-06 Bioconductor  
##  S4Vectors                         * 0.8.11   2016-01-29 Bioconductor  
##  stringi                             1.0-1    2015-10-22 CRAN (R 3.2.0)
##  stringr                             1.0.0    2015-04-30 CRAN (R 3.2.0)
##  SummarizedExperiment                1.0.2    2016-01-01 Bioconductor  
##  TxDb.Hsapiens.UCSC.hg19.knownGene * 3.2.2    2015-10-15 Bioconductor  
##  XML                                 3.98-1.3 2015-06-30 CRAN (R 3.2.0)
##  XVector                             0.10.0   2015-10-14 Bioconductor  
##  yaml                                2.1.13   2014-06-12 CRAN (R 3.2.0)
##  zlibbioc                            1.16.0   2015-10-14 Bioconductor</code></pre>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
